{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "vcf_file_address= \"22_2k.vcf\"#\"22_allps_200.vcf\"\n",
    "vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "\n",
    "first_het_variant_s1=True \n",
    "blockid_blocks_s1=[]\n",
    "allele_blocks_s1=[]\n",
    "varpos_blocks_s1=[]\n",
    "\n",
    "first_het_variant_s2=True \n",
    "blockid_blocks_s2=[]\n",
    "allele_blocks_s2=[]\n",
    "varpos_blocks_s2=[]\n",
    "var_blockids1_blocks_s2=[]   # for each var in s2, shows the s1 blockid\n",
    "\n",
    "header_lines_list=[]\n",
    "for line in vcf_file:\n",
    "    line_strip=line.strip()\n",
    "    if line_strip.startswith('#'):\n",
    "        header_lines_list.append(line_strip)\n",
    "        sample_names=line_strip.split('\\t')[9:] # last line of header contains sample name\n",
    "        #print(sample_names)\n",
    "    else:\n",
    "\n",
    "        line_parts=line_strip.split('\\t')\n",
    "        gt_flags, s1, s2, s3, s4, s5 = line_parts[8:]\n",
    "\n",
    "        varpos=int(line_parts[1]) # variant position\n",
    "        gt_flags_split=gt_flags.split(\":\")\n",
    "        \n",
    "        if \"PS\" in gt_flags_split:\n",
    "            idx_ps= gt_flags.split(\":\").index(\"PS\")\n",
    "            idx_gt= gt_flags.split(\":\").index(\"GT\")\n",
    "\n",
    "\n",
    "            split_s1=s1.split(\":\")\n",
    "            blockid_s1 = split_s1[idx_ps]\n",
    "            allele_s1=split_s1[idx_gt]\n",
    "            \n",
    "            if allele_s1==\"0/1\":allele_s1=\"./.\";blockid_s1=\".\"\n",
    "                \n",
    "            if not blockid_s1 == '.':\n",
    "                if first_het_variant_s1:           # for the first het variant\n",
    "                    first_het_variant_s1 = False\n",
    "                    \n",
    "                    allele_block_s1 = [int(allele_s1[0])]     # alleles of a block\n",
    "                    varpos_block_s1 = [varpos]               # var pos of a block\n",
    "                    blockid_blocks_s1.append(int(blockid_s1))   # add the blockid of the first var  to the list of blockids\n",
    "                    \n",
    "                else:                              # for the rest of het variants\n",
    "                    if int(blockid_s1) in blockid_blocks_s1:\n",
    "                        allele_block_s1.append(int(allele_s1[0]))\n",
    "                        varpos_block_s1.append(varpos)\n",
    "                        \n",
    "                    else: \n",
    "                        # add previous block to the list of all blocks\n",
    "                        allele_blocks_s1.append(allele_block_s1)\n",
    "                        varpos_blocks_s1.append(varpos_block_s1)\n",
    "\n",
    "                        # creat new phase block\n",
    "                        allele_block_s1 = [int(allele_s1[0])]\n",
    "                        varpos_block_s1 = [varpos]\n",
    "                        blockid_blocks_s1.append(int(blockid_s1))\n",
    "\n",
    "                        \n",
    "                        \n",
    "            split_s2=s2.split(\":\")\n",
    "            blockid_s2 = split_s2[idx_ps]\n",
    "            allele_s2=split_s2[idx_gt]\n",
    "            \n",
    "            if allele_s2==\"0/1\":allele_s2=\"./.\";blockid_s2=\".\"\n",
    "                \n",
    "            if not blockid_s2 == '.':\n",
    "                if first_het_variant_s2:           # for the first het variant\n",
    "                    first_het_variant_s2 = False\n",
    "                    \n",
    "                    allele_block_s2 = [int(allele_s2[0])]     # alleles of a block\n",
    "                    varpos_block_s2 = [varpos]               # var pos of a block\n",
    "                    blockid_blocks_s2.append(int(blockid_s2))   # add the blockid of the first var  to the list of blockids\n",
    "                    var_blockids1_block_s2=[blockid_s1]\n",
    "                    \n",
    "                else:                              # for the rest of het variants\n",
    "                    if int(blockid_s2) in blockid_blocks_s2:\n",
    "                        allele_block_s2.append(int(allele_s2[0]))\n",
    "                        varpos_block_s2.append(varpos)\n",
    "                        var_blockids1_block_s2.append(blockid_s1)\n",
    "                        \n",
    "                    else: \n",
    "                        # add previous block to the list of all blocks\n",
    "                        allele_blocks_s2.append(allele_block_s2)\n",
    "                        varpos_blocks_s2.append(varpos_block_s2)\n",
    "                        var_blockids1_blocks_s2.append(var_blockids1_block_s2)\n",
    "\n",
    "                        # creat new phase block\n",
    "                        allele_block_s2 = [int(allele_s2[0])]\n",
    "                        varpos_block_s2 = [varpos]\n",
    "                        blockid_blocks_s2.append(int(blockid_s2))\n",
    "                        var_blockids1_block_s2=[blockid_s1]\n",
    "                       \n",
    "                        \n",
    "                        \n",
    "# add last block to the list of all blocks\n",
    "allele_blocks_s1.append(allele_block_s1)\n",
    "varpos_blocks_s1.append(varpos_block_s1)\n",
    "\n",
    "allele_blocks_s2.append(allele_block_s2)\n",
    "varpos_blocks_s2.append(varpos_block_s2)\n",
    "var_blockids1_blocks_s2.append(var_blockids1_block_s2)\n",
    "                        \n",
    "    \n",
    "    \n",
    "number_phaseblocks_s2= len(allele_blocks_s2)\n",
    "    \n",
    "    \n",
    "\n",
    "    \n",
    "intersected_blocks_idins1_s2=[] # for each block in s2, the corresponding s1 block id\n",
    "\n",
    "for var_blockids1_blocks_s2_i in var_blockids1_blocks_s2:\n",
    "    var_blockids1_blocks_s2_i_set=set(var_blockids1_blocks_s2_i)\n",
    "    \n",
    "    if '.' in var_blockids1_blocks_s2_i_set:\n",
    "        var_blockids1_blocks_s2_i_set.discard('.')\n",
    "        \n",
    "    if len(var_blockids1_blocks_s2_i_set)==1:\n",
    "        intersected_idins1_s2=int(list(var_blockids1_blocks_s2_i_set)[0])\n",
    "    if len(var_blockids1_blocks_s2_i_set)==0:\n",
    "        intersected_idins1_s2=0\n",
    "    if len(var_blockids1_blocks_s2_i_set)>1:\n",
    "        # find the most frequent blockid in the phaseblock\n",
    "        intersected_idins1_s2=int(max(var_blockids1_blocks_s2_i_set, key=var_blockids1_blocks_s2_i.count) ) \n",
    "        \n",
    "    intersected_blocks_idins1_s2.append(intersected_idins1_s2)\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16065536, 16406342, 16657810, 16960113]\n",
      "22\n",
      "14\n",
      "12\n",
      "189\n",
      "[16065536, 16142229, 16406342, 16460985, 16657810, 16661781, 16960113, 16991976, 17092688, 17180682, 17273838]\n",
      "20\n",
      "2\n",
      "3\n",
      "11\n",
      "2\n",
      "11\n",
      "12\n",
      "45\n",
      "81\n",
      "16\n",
      "5\n"
     ]
    }
   ],
   "source": [
    "print(blockid_blocks_s1)\n",
    "for i in range(len(allele_blocks_s1)):\n",
    "    print(len(allele_blocks_s1[i]))\n",
    "    \n",
    "print(blockid_blocks_s2)\n",
    "for i in range(len(allele_blocks_s2)):\n",
    "    print(len(allele_blocks_s2[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "allele_blocks_s2_new=list(allele_blocks_s2)\n",
    "for block_i in range(number_phaseblocks_s2): # \n",
    "#for block_i, allele_block_s2 in enumerate(allele_blocks_s2):\n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "    intersected_idins1_s2=intersected_blocks_idins1_s2[block_i]\n",
    "    if intersected_idins1_s2 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s2)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s2)\n",
    "    similiar_count=0\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "        if allele_s1==allele_s2: similiar_count+=1\n",
    "    #print(similiar_count/len(intersect_varpos))\n",
    "    if similiar_count < len(intersect_varpos)/2:\n",
    "        #print(\"need\", similiar_count)\n",
    "        allele_block_s2_new= [1-i for i in allele_block_s2]\n",
    "        allele_blocks_s2_new[block_i]=allele_block_s2_new\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample2 0 16065536 16090396\n",
      "s1 16065536 \n",
      "\n",
      "sample2 1 16142229 16156069\n",
      "s1 16065536 \n",
      "\n",
      "sample2 2 16406342 16413241\n",
      "s1 16406342 \n",
      "\n",
      "sample2 3 16460985 16476596\n",
      "s1 16406342 \n",
      "\n",
      "sample2 4 16657810 16659787\n",
      "s1 16657810 \n",
      "\n",
      "sample2 5 16661781 16679716\n",
      "s1 16657810 \n",
      "\n",
      "sample2 6 16960113 16963253\n",
      "s1 16960113 \n",
      "\n",
      "sample2 7 16991976 17014571\n",
      "s1 16960113 \n",
      "\n",
      "sample2 8 17092688 17165240\n",
      "s1 16960113 \n",
      "\n",
      "sample2 9 17180682 17231422\n",
      "s1 16960113 \n",
      "\n",
      "sample2 10 17273838 17277238\n",
      "s1 16960113 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "block_i=0\n",
    "for block_i in range(len(allele_blocks_s2)):\n",
    "#for block_i, allele_block_s2 in enumerate(allele_blocks_s2):\n",
    "    check=0\n",
    "\n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "    start_pos_block=blockid_blocks_s2[block_i] # varpos_block_s2[0]\n",
    "    end_pos_block=varpos_block_s2[-1]\n",
    "    print(\"sample2\",block_i,start_pos_block ,end_pos_block)\n",
    "\n",
    "    for blockid_index_s1 in range(len(blockid_blocks_s1)):\n",
    "\n",
    "        blockid_s1_current=blockid_blocks_s1[blockid_index_s1]\n",
    "        \n",
    "        if blockid_index_s1 < len(blockid_blocks_s1)-1:\n",
    "            blockid_s1_next=blockid_blocks_s1[blockid_index_s1+1]\n",
    "        elif blockid_index_s1 == len(blockid_blocks_s1)-1:\n",
    "            blockid_s1_next= varpos_blocks_s1[-1][-1]\n",
    "            \n",
    "\n",
    "        # find the location of frist_block_s2 in s1_blocks\n",
    "        if start_pos_block >= blockid_s1_current and  start_pos_block < blockid_s1_next :\n",
    "            check=blockid_s1_current \n",
    "\n",
    "    print(\"s1\",check,'\\n')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample 1 0 16065536 16156069\n",
      "sample 1 1 16406342 16476596\n",
      "sample 1 2 16657810 16678960\n",
      "sample 1 3 16960113 17277238\n"
     ]
    }
   ],
   "source": [
    "for block_i in range(len(allele_blocks_s1)):\n",
    "#for block_i, allele_block_s2 in enumerate(allele_blocks_s2):\n",
    "\n",
    "\n",
    "    allele_block_s1=allele_blocks_s1[block_i]\n",
    "    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "    start_pos_block=blockid_blocks_s1[block_i] # varpos_block_s2[0]\n",
    "    end_pos_block=varpos_block_s1[-1]\n",
    "    print(\"sample 1\",block_i,start_pos_block ,end_pos_block)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
