{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16065536, '16065536']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[51242557],\n",
       " [16065774,\n",
       "  16065824,\n",
       "  16066157,\n",
       "  16067462,\n",
       "  16067500,\n",
       "  16067555,\n",
       "  16068688,\n",
       "  16069141,\n",
       "  16070517,\n",
       "  16071420,\n",
       "  16071624,\n",
       "  16072581,\n",
       "  16072799,\n",
       "  16073196,\n",
       "  16076439,\n",
       "  16080694,\n",
       "  16082335,\n",
       "  16089890,\n",
       "  16090396,\n",
       "  16142229,\n",
       "  16156069]]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(blockid_blocks_s1)\n",
    "varpos_blocks_s1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "import numpy as np\n",
    "# 22_ont_phased.vcf.gz 22_clr_phased.vcf.gz 22_hifi_phased.vcf.gz 22_ill_phased.vcf.gz 22_trio_true_ps.vcf.gz  > 22_all.vcf\n",
    "\n",
    "vcf_file_address= \"22_1k.vcf\"#\"22_allps_200.vcf\"\n",
    "\n",
    "\n",
    "matrix_address=\"matrix_\"+vcf_file_address\n",
    "\n",
    "matrix_file=open(matrix_address,'w');\n",
    "\n",
    "\n",
    "matrix_file.write(\"\\t\\t\".join([\"Pos\",\"ONT\",\"CLR\",\"HiFi\",\"Illm\",\"parental\"])+\"\\n\")\n",
    "\n",
    "\n",
    "\n",
    "sample1_blocks_pos={}\n",
    "sample1_blocks_allele={}\n",
    "\n",
    "sample1_block_pos=[]\n",
    "sample1_block_allele=[]\n",
    "\n",
    "\n",
    "\n",
    "if 1:\n",
    "#def  read_vcf_file_with_truth(vcf_file_address):\n",
    "    id_blocks_s1=[]\n",
    "    allele_blocks_s1=[]\n",
    "    var_pos_blocks_s1=[]\n",
    "    \n",
    "    \n",
    "    id_blocks_s2=[]\n",
    "    allele_blocks_s2=[]\n",
    "    var_pos_blocks_s2=[]\n",
    "    \n",
    "    \n",
    "    first_het_variant_s1=True\n",
    "    first_het_variant_s2=True\n",
    "\n",
    "    vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "    hap_blocks_sample1_dic={} # key: id of phase block (i.e. phase st), value= {genomic_position:allele_1}\n",
    "    hap_blocks_sample2_dic={}\n",
    "\n",
    "    hap_len_sample1_dic={}\n",
    "\n",
    "    header_lines_list=[]\n",
    "    for line in vcf_file:\n",
    "        line_strip=line.strip()\n",
    "        if line_strip.startswith('#'):\n",
    "            header_lines_list.append(line_strip)\n",
    "            sample_names=line_strip.split('\\t')[9:]# last line of header contains sample name\n",
    "            #print(sample_names)\n",
    "        else:\n",
    "\n",
    "            line_parts=line_strip.split('\\t')\n",
    "\n",
    "            gt_flags, sample1, sample2, sample3, sample4, sample5 = line_parts[8:]\n",
    "            #if \"|\" in sample1:\n",
    "\n",
    "            var_pos=int(line_parts[1])\n",
    "\n",
    "            gt_flags_split=gt_flags.split(\":\")\n",
    "            if \"PS\" in gt_flags_split:\n",
    "                idx_ps= gt_flags.split(\":\").index(\"PS\")\n",
    "                idx_gt= gt_flags.split(\":\").index(\"GT\")\n",
    "            \n",
    "                \n",
    "                sample1_split=sample1.split(\":\")\n",
    "                block_id1 = sample1_split[idx_ps] #int(\n",
    "                allele_sample1=sample1_split[idx_gt]\n",
    "                #if allele_sample1==\"0/1\":allele_sample1=\"./.\";block_id1=\".\"\n",
    "                \n",
    "                if not block_id1 == '.':\n",
    "                    if first_het_variant_s1:           # for the first het variant\n",
    "                        first_het_variant_s1 = False\n",
    "                        allele_block_s1 = [int(allele_sample1[0])]\n",
    "                        var_pos_block_s1 = [int(var_pos)]\n",
    "                        id_blocks_s1.append(block_id1)\n",
    "                    else:                              # for the rest of het variants\n",
    "                        if block_id1 in id_blocks:\n",
    "                            allele_block_s1.append(int(allele_sample1[0]))\n",
    "                            var_pos_block_s1.append(int(var_pos))\n",
    "                        else:\n",
    "\n",
    "                            # add previous block to the list of all blocks\n",
    "                            allele_blocks_s1.append(allele_block_s1)\n",
    "                            var_pos_blocks_s1.append(var_pos_block_s1)\n",
    "\n",
    "                            # creat new phase block\n",
    "                            allele_block_s1 = [int(allele_sample1[0])]\n",
    "                            var_pos_block_s1 = [int(var_pos)]\n",
    "                            id_blocks.append(block_id1)\n",
    "                \n",
    "\n",
    "                sample2_split=sample2.split(\":\")\n",
    "                block_id2 = sample2_split[idx_ps] #int(\n",
    "                allele_sample2=sample2_split[idx_gt]\n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "                if not block_id1 == '.':\n",
    "                    if first_het_variant_s2:           # for the first het variant\n",
    "                        first_het_variant_s2 = False\n",
    "                        allele_block_s2 = [int(allele_sample2[0])]\n",
    "                        var_pos_block_s2 = [int(var_pos)]\n",
    "                        id_blocks_s2.append(block_id2)\n",
    "                    else:                              # for the rest of het variants\n",
    "                        if block_id2 in id_blocks_s2:\n",
    "                            allele_block_s2.append(int(allele_sample2[0]))\n",
    "                            var_pos_block_s2.append(int(var_pos))\n",
    "                        else:\n",
    "\n",
    "                            # add previous block to the list of all blocks\n",
    "                            allele_blocks_s2.append(allele_block_s2)\n",
    "                            var_pos_blocks_s2.append(var_pos_block_s2)\n",
    "\n",
    "                            # creat new phase block\n",
    "                            allele_block_s2 = [int(allele_sample2[0])]\n",
    "                            var_pos_block_s2 = [int(var_pos)]\n",
    "                            id_blocks_s2.append(block_id2)\n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "                sample3_split=sample3.split(\":\")\n",
    "                block_id3 = sample3_split[idx_ps] #int(\n",
    "                allele_sample3=sample3_split[idx_gt]\n",
    "\n",
    "                sample4_split=sample4.split(\":\")\n",
    "                block_id4 = sample4_split[idx_ps] #int(\n",
    "                allele_sample4=sample4_split[idx_gt]\n",
    "\n",
    "                sample5_split=sample5.split(\":\")\n",
    "                block_id5 = sample5_split[idx_ps] #int(\n",
    "                allele_sample5=sample5_split[idx_gt]\n",
    "                \n",
    "                condition_phase=(\"|\" in allele_sample1) or (\"|\" in allele_sample2) or (\"|\" in allele_sample3) or (\"|\" in allele_sample4) or (\"|\" in allele_sample5)\n",
    "                \n",
    "                if (\"/\" in allele_sample1) and (\"/\" in allele_sample2) and (\"/\" in allele_sample3) and (\"/\" in allele_sample4):\n",
    "                    condition_phase= False\n",
    "                if block_id1=='.':block_id1=\".\\t\"\n",
    "                if block_id2=='.':block_id2=\".\\t\"\n",
    "                if block_id3=='.':block_id3=\".\\t\"\n",
    "                if block_id4=='.':block_id4=\".\\t\"\n",
    "                    \n",
    "                matrix_row_list=[str(var_pos),allele_sample1+\":\"+block_id1,allele_sample2+\":\"+block_id2,allele_sample3+\":\"+block_id3,\n",
    "                                allele_sample4+\":\"+block_id4,allele_sample5]                    \n",
    "                    \n",
    "                if condition_phase:\n",
    "                    matrix_file.write(\"\\t\".join(matrix_row_list)+\"\\n\")\n",
    "                #else:\n",
    "                #    print(matrix_row_list)\n",
    "                            # add previous block to the list of all blocks\n",
    "\n",
    "                    \n",
    "    # add last block to the list of all blocks\n",
    "    allele_blocks_s1.append(allele_block_s1)\n",
    "    var_pos_blocks_s1.append(var_pos_block_s1)\n",
    "\n",
    "    # add last block to the list of all blocks\n",
    "    allele_blocks_s2.append(allele_block_s2)\n",
    "    var_pos_blocks_s2.append(var_pos_block_s2)\n",
    "\n",
    "\n",
    "\n",
    "    matrix_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samp1 block 16065536\n",
      "samp2 block 16065536\n",
      "8 20 20\n",
      "Flip is needed.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "num_overlap=0\n",
    "num_same=0\n",
    "\n",
    "block_i_s2=0\n",
    "\n",
    "print(\"samp1 block\",id_blocks_s1[0])\n",
    "print(\"samp2 block\",id_blocks_s2[block_i_s2])\n",
    "\n",
    "\n",
    "for var_i, var_pos  in  enumerate(var_pos_blocks_s2[block_i_s2]): # one block\n",
    "    \n",
    "    allele_sample2= allele_blocks_s2[block_i_s2][var_i]\n",
    "\n",
    "    var_pos_blocks_s1_0 = var_pos_blocks_s1[0]\n",
    "    allele_blocks_s1_0 = allele_blocks_s1[0]\n",
    "    \n",
    "    if var_pos in var_pos_blocks_s1_0:\n",
    "        \n",
    "        num_overlap += 1\n",
    "        index_in_ref = var_pos_blocks_s1_0.index(var_pos)\n",
    "        allele_sample1= allele_blocks_s1_0[index_in_ref]\n",
    "        \n",
    "        if allele_sample2 == allele_sample1:\n",
    "            num_same += 1\n",
    "            \n",
    "\n",
    "print(num_same,num_overlap,len(var_pos_blocks_s2[0]))\n",
    "if num_same <num_overlap/2:\n",
    "    print(\"Flip is needed.\")\n",
    "else:\n",
    "    print(\"Flip is not needed.\")\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-43-ef0d985955f7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mallele_blocks_s2_0\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex_in_ref\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "\n",
    "allele_blocks_s2_0[index_in_ref]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['10x', 'CCS', 'CLR', 'ont', 'parental']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# 22_ont_phased.vcf.gz 22_clr_phased.vcf.gz 22_hifi_phased.vcf.gz 22_ill_phased.vcf.gz 22_trio_true_ps.vcf.gz  > 22_all.vcf\n",
    "\n",
    "vcf_file_address= \"22_all.vcf\"#\"22_allps_200.vcf\"\n",
    "\n",
    "\n",
    "matrix_address=\"matrix_one1_\"+vcf_file_address\n",
    "\n",
    "matrix_file=open(matrix_address,'w');\n",
    "\n",
    "\n",
    "matrix_file.write(\"\\t\\t\".join([\"Pos\",\"ONT\",\"CLR\",\"HiFi\",\"Illm\",\"parental\"])+\"\\n\")\n",
    "\n",
    "\n",
    "first_het_variant_s1=True \n",
    "id_blocks_s1=[]\n",
    "\n",
    "\n",
    "vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "hap_blocks_sample1_dic={} # key: id of phase block (i.e. phase st), value= {genomic_position:allele_1}\n",
    "hap_blocks_sample2_dic={}\n",
    "\n",
    "hap_len_sample1_dic={}\n",
    "\n",
    "header_lines_list=[]\n",
    "for line in vcf_file:\n",
    "    line_strip=line.strip()\n",
    "    if line_strip.startswith('#'):\n",
    "        header_lines_list.append(line_strip)\n",
    "        sample_names=line_strip.split('\\t')[9:]# last line of header contains sample name\n",
    "        #print(sample_names)\n",
    "    else:\n",
    "\n",
    "        line_parts=line_strip.split('\\t')\n",
    "\n",
    "        gt_flags, sample1, sample2, sample3, sample4, sample5 = line_parts[8:]\n",
    "\n",
    "        var_pos=int(line_parts[1])\n",
    "        gt_flags_split=gt_flags.split(\":\")\n",
    "        \n",
    "        if \"PS\" in gt_flags_split:\n",
    "            idx_ps= gt_flags.split(\":\").index(\"PS\")\n",
    "            idx_gt= gt_flags.split(\":\").index(\"GT\")\n",
    "\n",
    "\n",
    "            sample1_split=sample1.split(\":\")\n",
    "            block_id1 = sample1_split[idx_ps] #int(\n",
    "            allele_sample1=sample1_split[idx_gt]\n",
    "            if allele_sample1==\"0/1\":allele_sample1=\"./.\";block_id1=\".\"\n",
    "\n",
    "\n",
    "            #if allele_sample1==\"0/1\":allele_sample1=\"./.\";block_id1=\".\"\n",
    "\n",
    "            if not block_id1 == '.':\n",
    "                if first_het_variant_s1:           # for the first het variant\n",
    "                    first_het_variant_s1 = False\n",
    "                    allele_block_s1 = [int(allele_sample1[0])]\n",
    "                    var_pos_block_s1 = [var_pos]\n",
    "                    id_blocks_s1.append(int(block_id1))\n",
    "                else:                              # for the rest of het variants\n",
    "                    if block_id1 in id_blocks_s1:\n",
    "                        allele_block_s1.append(int(allele_sample1[0]))\n",
    "                        var_pos_block_s1.append(int(var_pos))\n",
    "                    else:\n",
    "\n",
    "                        # add previous block to the list of all blocks\n",
    "                        allele_blocks_s1.append(allele_block_s1)\n",
    "                        var_pos_blocks_s1.append(var_pos_block_s1)\n",
    "\n",
    "                        # creat new phase block\n",
    "                        allele_block_s1 = [int(allele_sample1[0])]\n",
    "                        var_pos_block_s1 = [int(var_pos)]\n",
    "                        id_blocks.append(block_id1)\n",
    "\n",
    "\n",
    "\n",
    "            sample2_split=sample2.split(\":\")\n",
    "            block_id2 = sample2_split[idx_ps] #int(\n",
    "            allele_sample2=sample2_split[idx_gt]\n",
    "\n",
    "            sample3_split=sample3.split(\":\")\n",
    "            block_id3 = sample3_split[idx_ps] #int(\n",
    "            allele_sample3=sample3_split[idx_gt]\n",
    "\n",
    "            sample4_split=sample4.split(\":\")\n",
    "            block_id4 = sample4_split[idx_ps] #int(\n",
    "            allele_sample4=sample4_split[idx_gt]\n",
    "\n",
    "            sample5_split=sample5.split(\":\")\n",
    "            block_id5 = sample5_split[idx_ps] #int(\n",
    "            allele_sample5=sample5_split[idx_gt]\n",
    "\n",
    "            condition_phase=(\"|\" in allele_sample1) or (\"|\" in allele_sample2) or (\"|\" in allele_sample3) or (\"|\" in allele_sample4) or (\"|\" in allele_sample5)\n",
    "\n",
    "            if (\"/\" in allele_sample1) and (\"/\" in allele_sample2) and (\"/\" in allele_sample3) and (\"/\" in allele_sample4):\n",
    "                condition_phase= False\n",
    "            if block_id1=='.':block_id1=\".\\t\";allele_sample1='.'\n",
    "            if block_id2=='.':block_id2=\".\\t\";allele_sample2='.'\n",
    "            if block_id3=='.':block_id3=\".\\t\";allele_sample3='.'\n",
    "            if block_id4=='.':block_id4=\".\\t\";allele_sample4='.'\n",
    "\n",
    "            if allele_sample1[0]=='.':\n",
    "                new1=allele_sample1\n",
    "            else:\n",
    "                new1=str(1-int(allele_sample1[0]))\n",
    "\n",
    "            if allele_sample5[0]=='.':\n",
    "                new5=allele_sample5\n",
    "            else:\n",
    "                new5=str(int(allele_sample5[0]))\n",
    "\n",
    "\n",
    "            if allele_sample3[0]=='.':\n",
    "                new3=allele_sample3\n",
    "            else:\n",
    "                new3=str(1-int(allele_sample3[0]))\n",
    "\n",
    "\n",
    "            matrix_row_list=[str(var_pos),new1+\":\"+block_id1,allele_sample2[0]+\":\"+block_id2,new3+\":\"+block_id3,\n",
    "                            allele_sample4[0]+\":\"+block_id4,new5[0]]                    \n",
    "\n",
    "            if condition_phase:\n",
    "                matrix_file.write(\"\\t\".join(matrix_row_list)+\"\\n\")\n",
    "            #else:\n",
    "            #    print(matrix_row_list)\n",
    "\n",
    "    matrix_file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
