{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num of phaseblocks in sample one is  4\n",
      "Num of phaseblocks in sample two is  128\n",
      "number of occurance of weird=two variants are in the same block, but far from each other  0\n"
     ]
    }
   ],
   "source": [
    "\n",
    "vcf_file_address =  \"a/22_2k.vcf\"# \"22_all.vcf\"  #\"22_allps_200.vcf\"  \n",
    "vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "\n",
    "blockid_blocks_s1 = []\n",
    "allele_blocks_s1 = []\n",
    "varpos_blocks_s1 = []\n",
    "\n",
    "\n",
    "blockid_blocks_s2 = []\n",
    "allele_blocks_s2 = []\n",
    "varpos_blocks_s2 = []\n",
    "\n",
    "var_blockids1_blocks_s2 = []   # for each var in s2, shows the s1 blockid\n",
    "\n",
    "# for parental that it's only one block per chr so it doesn't have blockid\n",
    "alleles_s5 = []\n",
    "varposs_s5 = []\n",
    "\n",
    "\n",
    "varposs_all = []\n",
    "\n",
    "num_weird_occurs=0\n",
    "\n",
    "header_lines_list = []\n",
    "for line in vcf_file:\n",
    "    line_strip = line.strip()\n",
    "    \n",
    "    if line_strip.startswith('#'):\n",
    "        header_lines_list.append(line_strip)\n",
    "        sample_names = line_strip.split('\\t')[9:]   # last line of header contains sample name\n",
    "        #print(sample_names)\n",
    "        \n",
    "    else:\n",
    "        line_parts = line_strip.split('\\t')\n",
    "        \n",
    "        #gt_flags, s1, sss, ss, s2, s5 = line_parts[8:]\n",
    "        gt_flags, s1, s2, s5 = line_parts[8:]       # hifi illumina true\n",
    "\n",
    "        varpos = int(line_parts[1])                 # variant position\n",
    "        \n",
    "        varposs_all.append(varpos)\n",
    "        gt_flags_split = gt_flags.split(\":\")\n",
    "        \n",
    "        if \"PS\" in gt_flags_split:\n",
    "            idx_ps = gt_flags.split(\":\").index(\"PS\")\n",
    "            idx_gt = gt_flags.split(\":\").index(\"GT\")\n",
    "            \n",
    "\n",
    "                                                 # parental - last sample in the vcf file\n",
    "            split_s5 = s5.split(\":\")\n",
    "            allele_s5 = split_s5[idx_gt]\n",
    "            blockid_s5 = split_s5[idx_ps]\n",
    "            \n",
    "            if allele_s5 != \"./.\":\n",
    "                alleles_s5.append(int(allele_s5[0]))\n",
    "                varposs_s5.append(varpos)\n",
    "\n",
    "            \n",
    "            \n",
    "                                                ###  sample 1 ###\n",
    "            split_s1 = s1.split(\":\")\n",
    "            blockid_s1 = split_s1[idx_ps]\n",
    "            allele_s1 = split_s1[idx_gt]\n",
    "            \n",
    "\n",
    "            if blockid_s1 == \".\":                                 # unphased variants for sample one\n",
    "                pass\n",
    "            \n",
    "            elif blockid_s1 in blockid_blocks_s1:\n",
    "                assert allele_s1[1] == \"|\", \"allele_s1[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                #if blockid_s1 == previous_blockid_s1: previous_blockid_s1_idx has been defined in previous variant which is the last idx\n",
    "                if blockid_s1 != previous_blockid_s1:  \n",
    "                    # weird situation happens for whatshap in which two variants are in the same block but in between there is/are other phaseblock(s)\n",
    "                    # note that the unphased variants are ignored sooner and are not considered here\n",
    "                    previous_blockid_s1_idx = blockid_blocks_s1.index(blockid_s1)\n",
    "                    #print(\"two variants are in the same block, but far from each other\",varpos)                \n",
    "                    num_weird_occurs+=1\n",
    "\n",
    "                allele_blocks_s1[previous_blockid_s1_idx].append( int(allele_s1[0]) )\n",
    "                varpos_blocks_s1[previous_blockid_s1_idx].append( varpos ) \n",
    "\n",
    "                \n",
    "            else:                                                  # A new block started, it can be the first block\n",
    "                \n",
    "                allele_blocks_s1.append( [int(allele_s1[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s1.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s1.append( blockid_s1 )             # add the blockid of the first var  to the list of blockids\n",
    "                \n",
    "                previous_blockid_s1 = blockid_s1\n",
    "                previous_blockid_s1_idx = len(blockid_blocks_s1)-1  # the index of the last block id\n",
    "                \n",
    "\n",
    "\n",
    "                \n",
    "       \n",
    "    \n",
    "                                                ###  sample 2 ###\n",
    "            split_s2 = s2.split(\":\")\n",
    "            blockid_s2 = split_s2[idx_ps]\n",
    "            allele_s2 = split_s2[idx_gt]\n",
    "            \n",
    "\n",
    "            if blockid_s2 == \".\":\n",
    "                pass\n",
    "            \n",
    "            elif blockid_s2 in blockid_blocks_s2:\n",
    "                assert allele_s2[1] == \"|\", \"allele_s2[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                #if blockid_s2 == previous_blockid_s2: previous_blockid_s2_idx has been defined in previous variant\n",
    "                if blockid_s2 != previous_blockid_s2:  \n",
    "                    # weird situation happens for whatshap in which two variants are in the same block but in between there is/are other phaseblock(s)\n",
    "                    # note that the unphased variants are ignored sooner and are not considered here\n",
    "                    previous_blockid_s2_idx = blockid_blocks_s2.index(blockid_s2)\n",
    "                    #print(\"two variants are in the same block, but far from each other\",varpos)\n",
    "                    num_weird_occurs+=1\n",
    "\n",
    "                allele_blocks_s2[previous_blockid_s2_idx].append( int(allele_s2[0]) )\n",
    "                varpos_blocks_s2[previous_blockid_s2_idx].append( varpos ) \n",
    "                var_blockids1_blocks_s2[previous_blockid_s2_idx].append( blockid_s1 )\n",
    "                \n",
    "            else:                                                 # A new block started   (it can be the first block)\n",
    "                \n",
    "                allele_blocks_s2.append( [int(allele_s2[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s2.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s2.append( blockid_s2 )             # add the blockid of the first var  to the list of blockids\n",
    "                \n",
    "                var_blockids1_blocks_s2.append( [blockid_s1] )  \n",
    "                \n",
    "                previous_blockid_s2 = blockid_s2\n",
    "                previous_blockid_s2_idx = len(blockid_blocks_s2)-1   # the index of the last block id\n",
    "                \n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "number_phaseblocks_s1= len(allele_blocks_s1)\n",
    "print(\"Num of phaseblocks in sample one is \",number_phaseblocks_s1)\n",
    " \n",
    "number_phaseblocks_s2= len(allele_blocks_s2)\n",
    "print(\"Num of phaseblocks in sample two is \",number_phaseblocks_s2)\n",
    " \n",
    "print(\"number of occurance of weird=two variants are in the same block, but far from each other \",num_weird_occurs)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test code\n",
    "# cd /Volumes/work/myjupyter/phaseyou\n",
    "#grep \"46014176\" 22_all.vcf\n",
    "\n",
    "#blockid_blocks_s2.index(\"46014176\")\n",
    "#print(allele_blocks_s2[5568],\"\\n\",varpos_blocks_s2[5568])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# vcf_file_address= \"22_all.vcf\"  #\"22_allps_200.vcf\"\n",
    "# vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "\n",
    "# first_het_variant_s1=True \n",
    "# blockid_blocks_s1=[]\n",
    "# allele_blocks_s1=[]\n",
    "# varpos_blocks_s1=[]\n",
    "\n",
    "# first_het_variant_s2=True \n",
    "# blockid_blocks_s2=[]\n",
    "# allele_blocks_s2=[]\n",
    "# varpos_blocks_s2=[]\n",
    "# var_blockids1_blocks_s2=[]   # for each var in s2, shows the s1 blockid\n",
    "\n",
    "\n",
    "# alleles_s5=[]\n",
    "# varposs_s5=[]\n",
    "\n",
    "# header_lines_list=[]\n",
    "# for line in vcf_file:\n",
    "#     line_strip=line.strip()\n",
    "#     if line_strip.startswith('#'):\n",
    "#         header_lines_list.append(line_strip)\n",
    "#         sample_names=line_strip.split('\\t')[9:] # last line of header contains sample name\n",
    "#         #print(sample_names)\n",
    "#     else:\n",
    "\n",
    "#         line_parts=line_strip.split('\\t')\n",
    "#         #gt_flags, s1, sss, ss, s2, s5 = line_parts[8:]\n",
    "#         gt_flags, s1, s2, s5 = line_parts[8:] # hifi illumina true\n",
    "\n",
    "#         varpos=int(line_parts[1]) # variant position\n",
    "#         gt_flags_split=gt_flags.split(\":\")\n",
    "        \n",
    "#         if \"PS\" in gt_flags_split:\n",
    "#             idx_ps= gt_flags.split(\":\").index(\"PS\")\n",
    "#             idx_gt= gt_flags.split(\":\").index(\"GT\")\n",
    "            \n",
    "            \n",
    "#             # parental\n",
    "#             split_s5 = s5.split(\":\")\n",
    "#             blockid_s5 = split_s5[idx_ps]\n",
    "#             allele_s5 = split_s5[idx_gt]\n",
    "#             if allele_s5 !=\"./.\":\n",
    "#                 alleles_s5.append(int(allele_s5[0]))\n",
    "#                 varposs_s5.append(varpos)\n",
    "\n",
    "            \n",
    "#             split_s1=s1.split(\":\")\n",
    "#             blockid_s1 = split_s1[idx_ps]\n",
    "#             allele_s1=split_s1[idx_gt]\n",
    "            \n",
    "#             #if allele_s1==\"0/1\":allele_s1=\"./.\";blockid_s1=\".\"\n",
    "#             if allele_s1!=\"0|1\"  or allele_s1!=\"1|0\" : allele_s1=\"8/8\";blockid_s1=9\n",
    "                \n",
    "#             if 1: #not blockid_s1 == '.':\n",
    "#                 if first_het_variant_s1:           # for the first het variant\n",
    "#                     first_het_variant_s1 = False\n",
    "                    \n",
    "#                     allele_block_s1 = [int(allele_s1[0])]     # alleles of a block\n",
    "#                     varpos_block_s1 = [varpos]               # var pos of a block\n",
    "#                     blockid_blocks_s1.append(int(blockid_s1))   # add the blockid of the first var  to the list of blockids\n",
    "                    \n",
    "#                 else:                              # for the rest of het variants\n",
    "\n",
    "#                     if int(blockid_s1) in blockid_blocks_s1:\n",
    "#                         allele_block_s1.append(int(allele_s1[0]))\n",
    "#                         varpos_block_s1.append(varpos)\n",
    "                        \n",
    "#                     else: \n",
    "#                         # add previous block to the list of all blocks\n",
    "#                         allele_blocks_s1.append(allele_block_s1)\n",
    "#                         varpos_blocks_s1.append(varpos_block_s1)\n",
    "\n",
    "#                         # creat new phase block\n",
    "#                         allele_block_s1 = [int(allele_s1[0])]\n",
    "#                         varpos_block_s1 = [varpos]\n",
    "#                         blockid_blocks_s1.append(int(blockid_s1))\n",
    "\n",
    "                        \n",
    "            \n",
    "#             split_s2=s2.split(\":\")\n",
    "#             blockid_s2 = split_s2[idx_ps]\n",
    "#             allele_s2=split_s2[idx_gt]\n",
    "            \n",
    "#             #if allele_s2==\"0/1\":allele_s2=\"./.\";blockid_s2=\".\"\n",
    "#             if allele_s2!=\"0|1\"  or allele_s2!=\"1|0\": allele_s2=\"8/8\";blockid_s2=9\n",
    "                \n",
    "#             if 1: #not blockid_s2 == '.':\n",
    "#                 if first_het_variant_s2:           # for the first het variant\n",
    "#                     first_het_variant_s2 = False\n",
    "                    \n",
    "#                     allele_block_s2 = [int(allele_s2[0])]     # alleles of a block\n",
    "#                     varpos_block_s2 = [varpos]               # var pos of a block\n",
    "#                     blockid_blocks_s2.append(int(blockid_s2))   # add the blockid of the first var  to the list of blockids\n",
    "#                     var_blockids1_block_s2=[blockid_s1]\n",
    "                    \n",
    "#                 else:                              # for the rest of het variants\n",
    "#                     if int(blockid_s2) in blockid_blocks_s2:\n",
    "#                         allele_block_s2.append(int(allele_s2[0]))\n",
    "#                         varpos_block_s2.append(varpos)\n",
    "#                         var_blockids1_block_s2.append(blockid_s1)\n",
    "                        \n",
    "#                     else: \n",
    "#                         # add previous block to the list of all blocks\n",
    "#                         allele_blocks_s2.append(allele_block_s2)\n",
    "#                         varpos_blocks_s2.append(varpos_block_s2)\n",
    "#                         var_blockids1_blocks_s2.append(var_blockids1_block_s2)\n",
    "\n",
    "#                         # creat new phase block\n",
    "#                         allele_block_s2 = [int(allele_s2[0])]\n",
    "#                         varpos_block_s2 = [varpos]\n",
    "#                         blockid_blocks_s2.append(int(blockid_s2))\n",
    "#                         var_blockids1_block_s2=[blockid_s1]\n",
    "                       \n",
    "                        \n",
    "                        \n",
    "# # add last block to the list of all blocks\n",
    "# allele_blocks_s1.append(allele_block_s1)\n",
    "# varpos_blocks_s1.append(varpos_block_s1)\n",
    "\n",
    "# allele_blocks_s2.append(allele_block_s2)\n",
    "# varpos_blocks_s2.append(varpos_block_s2)\n",
    "# var_blockids1_blocks_s2.append(var_blockids1_block_s2)\n",
    "                        \n",
    "    \n",
    "    \n",
    "# number_phaseblocks_s2= len(allele_blocks_s2)\n",
    "# print(\"num s2:\",number_phaseblocks_s2)\n",
    "\n",
    "# number_phaseblocks_s1= len(allele_blocks_s1)\n",
    "# print(\"num s1:\",number_phaseblocks_s1)\n",
    " \n",
    "    \n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1087\n"
     ]
    }
   ],
   "source": [
    "#%pwd\n",
    "num_phased_s1=0\n",
    "for varpos_block_s1 in varpos_blocks_s1:\n",
    "    num_phased_s1+=len(varpos_block_s1)\n",
    "    if len(varpos_block_s1)<2:\n",
    "        print(\"small\",varpos_block_s1)\n",
    "print(num_phased_s1)\n",
    "\n",
    "\n",
    "# num_phased_s2=0\n",
    "# for allele_block_s2 in allele_blocks_s2:\n",
    "#     num_phased_s2+=len(allele_block_s2)\n",
    "#     if len(allele_block_s2)<2:\n",
    "#         print(\"small,s2\",len(allele_block_s2))\n",
    "# print(num_phased_s2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1083\n",
      "['16050159', '16050159']\n"
     ]
    }
   ],
   "source": [
    "num_phased_a=0\n",
    "for a in var_blockids1_blocks_s2:\n",
    "    num_phased_a+=len(a)\n",
    "print(num_phased_a)   \n",
    "#len(var_blockids1_blocks_s2[100])\n",
    "print(var_blockids1_blocks_s2[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_blocks_idins1_s2=[] # for each block in s2, the corresponding s1 block id\n",
    "\n",
    "for var_blockids1_blocks_s2_i in var_blockids1_blocks_s2:\n",
    "    var_blockids1_blocks_s2_i_set=set(var_blockids1_blocks_s2_i)\n",
    "    \n",
    "    if '.' in var_blockids1_blocks_s2_i_set:\n",
    "        var_blockids1_blocks_s2_i_set.discard('.')\n",
    "        \n",
    "    if len(var_blockids1_blocks_s2_i_set)==1:\n",
    "        intersected_idins1_s2=list(var_blockids1_blocks_s2_i_set)[0]\n",
    "    if len(var_blockids1_blocks_s2_i_set)==0:\n",
    "        intersected_idins1_s2=0\n",
    "    if len(var_blockids1_blocks_s2_i_set)>1:\n",
    "        # find the most frequent blockid in the phaseblock\n",
    "        intersected_idins1_s2=max(var_blockids1_blocks_s2_i_set, key=var_blockids1_blocks_s2_i.count) \n",
    "    \n",
    "        \n",
    "    intersected_blocks_idins1_s2.append(intersected_idins1_s2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(blockid_blocks_s1)\n",
    "#for i in range(len(allele_blocks_s1)):\n",
    "#    print(len(allele_blocks_s1[i]))\n",
    "    \n",
    "#print(blockid_blocks_s2)\n",
    "#for i in range(len(allele_blocks_s2)):\n",
    "#    print(len(allele_blocks_s2[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12 7\n",
      "23 16\n",
      "39 35\n",
      "24 5\n"
     ]
    }
   ],
   "source": [
    "# re-ordering  the phase block allel based on parent\n",
    "\n",
    "number_phaseblocks_s1= len(allele_blocks_s1)\n",
    "\n",
    "#allele_blocks_s1_new=list(allele_blocks_s1)\n",
    "for block_i in range(number_phaseblocks_s1): # \n",
    "   \n",
    "    allele_block_s1=allele_blocks_s1[block_i]\n",
    "    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "\n",
    "    similiar_count=0\n",
    "    diss_count=0\n",
    "    for var_i, varpos in enumerate(varpos_block_s1):\n",
    "        allele_s1= allele_block_s1[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "            if allele_s1 == allele_s5:\n",
    "                similiar_count+=1\n",
    "            else: \n",
    "                diss_count+=1\n",
    "                \n",
    "    print(similiar_count, diss_count)          \n",
    "    if similiar_count < diss_count:\n",
    "        allele_block_s1_reordered = [1-i for i in allele_block_s1] # converting 0->1 and 1->0\n",
    "        allele_blocks_s1[block_i]=allele_block_s1_reordered\n",
    "        print(\"it needs to re-order\")\n",
    "    \n",
    "#         allele_block_s1_new = [1-i for i in allele_block_s1]\n",
    "#         allele_blocks_s1_new[block_i]=allele_block_s1_new\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#allele_blocks_s2_new=list(allele_blocks_s2)\n",
    "for block_i in range(number_phaseblocks_s2): # \n",
    "#for block_i, allele_block_s2 in enumerate(allele_blocks_s2):\n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "    intersected_idins1_s2=intersected_blocks_idins1_s2[block_i]\n",
    "    if intersected_idins1_s2 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s2)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    #allele_block_s1=allele_blocks_s1_new[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s2)\n",
    "    similiar_count=0\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "        if allele_s1==allele_s2: similiar_count+=1\n",
    "    #print(similiar_count/len(intersect_varpos))\n",
    "    if similiar_count < len(intersect_varpos)/2:\n",
    "        #print(\"need\", similiar_count)\n",
    "        allele_block_s2_reordered= [1-i for i in allele_block_s2]\n",
    "        allele_blocks_s2[block_i]=allele_block_s2_reordered\n",
    "        print(\"it needs to re-order\")\n",
    "#         allele_block_s2_new= [1-i for i in allele_block_s2]\n",
    "#         allele_blocks_s2_new[block_i]=allele_block_s2_new\n",
    "\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(blockid_blocks_s1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# matrix_address=\"matrix_up2_\"+vcf_file_address+\".txt\"\n",
    "\n",
    "# matrix_file=open(matrix_address,'w');\n",
    "# matrix_file.write(\"\\t\\t\".join([\"Pos\",\"HiFi\",\"Illm\",\"parental\"])+\"\\n\") # \"Pos\",\"ONT\",\"CLR\",\"HiFi\",\"Illm\",\"parental\"\n",
    "\n",
    "\n",
    "# for block_i_s1 in range(len(blockid_blocks_s1)):\n",
    "#     blockid_block_s1=blockid_blocks_s1[block_i_s1]\n",
    "    \n",
    "#     allele_block_s1=allele_blocks_s1_new[block_i_s1]\n",
    "#     varpos_block_s1=varpos_blocks_s1[block_i_s1]\n",
    "#     for var_i_s1 in range(len(varpos_block_s1)):\n",
    "#         varpos=varpos_block_s1[var_i_s1]\n",
    "#         allele_s1=allele_block_s1[var_i_s1]\n",
    "        \n",
    "#         allele_s2='.'\n",
    "#         blockid_s2='.'\n",
    "#         for block_i_s2 in range(len(blockid_blocks_s2)):\n",
    "#             blockid_block_s2 = blockid_blocks_s2[block_i_s2]\n",
    "#             allele_block_s2 = allele_blocks_s2_new[block_i_s2]\n",
    "#             varpos_block_s2 = varpos_blocks_s2[block_i_s2]\n",
    "            \n",
    "#             if varpos in varpos_block_s2:\n",
    "#                 allele_s2 = allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "#                 blockid_s2=blockid_block_s2\n",
    "#                 break # when var is found, no need to continue for loop\n",
    "            \n",
    "#         if varpos in varposs_s5:\n",
    "#             allele_s5 = alleles_s5[varposs_s5.index(varpos)]\n",
    "#         else: allele_s5= '.'\n",
    "\n",
    "                \n",
    "#         matrix_row_list=[str(varpos),str(allele_s1)+\":\"+str(blockid_block_s1),\n",
    "#                          str(allele_s2)+\":\"+str(blockid_s2),str(allele_s5)] \n",
    "#         matrix_file.write(\"\\t\".join(matrix_row_list)+\"\\n\")\n",
    "    \n",
    "# matrix_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix_address=vcf_file_address+\"_matrix.txt\"\n",
    "\n",
    "matrix_file=open(matrix_address,'w');\n",
    "matrix_file.write(\"\\t\\t\".join([\"Pos\",\"HiFi\",\"Illm\",\"parental\"])+\"\\n\") # \"Pos\",\"ONT\",\"CLR\",\"HiFi\",\"Illm\",\"parental\"\n",
    "\n",
    "for varpos in varposs_all:\n",
    "    # for increasing the speed we can use the information that the blocks are in order.\n",
    "    \n",
    "    \n",
    "    allele_s1='.'\n",
    "    blockid_s1='.'\n",
    "    for block_i_s1 in range(len(blockid_blocks_s1)):\n",
    "        allele_block_s1 = allele_blocks_s1[block_i_s1]\n",
    "        varpos_block_s1 = varpos_blocks_s1[block_i_s1]\n",
    "\n",
    "        if varpos in varpos_block_s1:\n",
    "            allele_s1 = allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "            blockid_s1 = blockid_blocks_s1[block_i_s1]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "            \n",
    "    allele_s2='.'\n",
    "    blockid_s2='.'\n",
    "    for block_i_s2 in range(len(blockid_blocks_s2)):\n",
    "        allele_block_s2 = allele_blocks_s2[block_i_s2]\n",
    "        varpos_block_s2 = varpos_blocks_s2[block_i_s2]\n",
    "\n",
    "        if varpos in varpos_block_s2:\n",
    "            allele_s2 = allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "            blockid_s2 = blockid_blocks_s2[block_i_s2]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "    \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5 = alleles_s5[varposs_s5.index(varpos)]\n",
    "        else: allele_s5= '.'\n",
    "\n",
    "        if not (allele_s1=='.' and allele_s2=='.' and allele_s5=='.'): \n",
    "            matrix_row_list=[str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "                             str(allele_s2)+\":\"+str(blockid_s2),str(allele_s5)] \n",
    "            matrix_file.write(\"\\t\".join(matrix_row_list)+\"\\n\")\n",
    "\n",
    "matrix_file.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16069374, 0, 0, 1] \n",
      "\n",
      "[16080579, 0, 0, 1] \n",
      "\n",
      "[16086302, 0, 0, 1] \n",
      "\n",
      "[16087217, 0, 0, 1] \n",
      "\n",
      "[16101684, 1, 1, 0] \n",
      "\n",
      "[16104385, 1, 1, 0] \n",
      "\n",
      "[16105717, 1, 0, 0] \n",
      "\n",
      "[16113547, 0, 0, 1] \n",
      "\n",
      "[16117006, 0, 1, 1] \n",
      "\n",
      "[16120074, 1, 0, 0] \n",
      "\n",
      "[16123812, 1, 0, 0] \n",
      "\n",
      "[16137152, 1, 0, 0] \n",
      "\n",
      "[16142002, 1, 1, 0] \n",
      "\n",
      "[16143370, 1, 1, 0] \n",
      "\n",
      "[16143995, 1, 1, 0] \n",
      "\n",
      "[16144768, 1, 1, 0] \n",
      "\n",
      "[16144952, 1, 1, 0] \n",
      "\n",
      "[16149963, 1, 0, 0] \n",
      "\n",
      "[16159540, 0, 0, 1] \n",
      "\n",
      "[16164946, 1, 1, 0] \n",
      "\n",
      "[16183780, 1, 1, 0] \n",
      "\n",
      "[16183790, 1, 1, 0] \n",
      "\n",
      "[16194681, 1, 0, 0] \n",
      "\n",
      "[16199059, 1, 0, 0] \n",
      "\n",
      "[16201114, 1, 0, 0] \n",
      "\n",
      "[16201332, 1, 0, 0] \n",
      "\n",
      "[16201360, 1, 0, 0] \n",
      "\n",
      "[16203241, 1, 1, 0] \n",
      "\n",
      "[16203542, 0, 1, 1] \n",
      "\n",
      "[16204790, 1, 0, 0] \n",
      "\n",
      "[16204824, 1, 0, 0] \n",
      "\n",
      "[16207034, 1, 0, 0] \n",
      "\n",
      "[16212603, 1, 0, 0] \n",
      "\n",
      "[16213011, 1, 0, 0] \n",
      "\n",
      "[16214176, 1, 1, 0] \n",
      "\n",
      "[16216383, 1, 1, 0] \n",
      "\n",
      "[16217414, 1, 1, 0] \n",
      "\n",
      "[16217477, 1, 1, 0] \n",
      "\n",
      "[16219445, 1, 0, 0] \n",
      "\n",
      "[16223201, 1, 0, 0] \n",
      "\n",
      "[16228164, 1, 1, 0] \n",
      "\n",
      "[16228270, 1, 1, 0] \n",
      "\n",
      "[16230082, 1, 1, 0] \n",
      "\n",
      "[16230525, 1, 1, 0] \n",
      "\n",
      "[16231828, 1, 0, 0] \n",
      "\n",
      "[16240092, 1, 1, 0] \n",
      "\n",
      "[16246137, 0, 0, 1] \n",
      "\n",
      "[16258516, 1, 1, 0] \n",
      "\n",
      "[16267436, 1, 1, 0] \n",
      "\n",
      "[16296459, 0, 0, 1] \n",
      "\n",
      "[16317765, 0, 1, 1] \n",
      "\n",
      "[16335320, 1, 1, 0] \n",
      "\n",
      "[16335338, 1, 1, 0] \n",
      "\n",
      "[16337995, 1, 1, 0] \n",
      "\n",
      "[16360538, 1, 0, 0] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "diff_list=[]\n",
    "\n",
    "for block_i in range(len(allele_blocks_s1)): # \n",
    "   \n",
    "    allele_block_s1=allele_blocks_s1[block_i]\n",
    "    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "\n",
    "    for var_i, varpos in enumerate(varpos_block_s1):\n",
    "        allele_s1= allele_block_s1[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "        \n",
    "            if allele_s1!=allele_s5:\n",
    "                \n",
    "                for block_i in range(number_phaseblocks_s2): # \n",
    "   \n",
    "                    allele_block_s2=allele_blocks_s2[block_i]\n",
    "                    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "                    if varpos in varpos_block_s2:\n",
    "                        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "                        diff_list.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                        print([varpos,allele_s1,allele_s2,allele_s5],\"\\n\")\n",
    "                        break\n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[16069374, 0, 0, 1] \n",
      "\n",
      "[16073196, 0, 0, 1] \n",
      "\n",
      "[16080579, 0, 0, 1] \n",
      "\n",
      "[16086302, 0, 0, 1] \n",
      "\n",
      "[16087217, 0, 0, 1] \n",
      "\n",
      "[16101684, 1, 1, 0] \n",
      "\n",
      "[16104385, 0, 1, 0] \n",
      "\n",
      "[16105128, 1, 0, 1] \n",
      "\n",
      "[16113547, 1, 0, 1] \n",
      "\n",
      "[16127345, 1, 0, 1] \n",
      "\n",
      "[16138401, 1, 1, 0] \n",
      "\n",
      "[16142002, 1, 1, 0] \n",
      "\n",
      "[16142229, 1, 1, 0] \n",
      "\n",
      "[16142899, 1, 1, 0] \n",
      "\n",
      "[16143370, 0, 1, 0] \n",
      "\n",
      "[16143995, 1, 1, 0] \n",
      "\n",
      "[16144768, 1, 1, 0] \n",
      "\n",
      "[16144952, 1, 1, 0] \n",
      "\n",
      "[16147765, 0, 1, 0] \n",
      "\n",
      "[16150751, 1, 0, 1] \n",
      "\n",
      "[16159540, 1, 0, 1] \n",
      "\n",
      "[16160468, 1, 0, 1] \n",
      "\n",
      "[16164946, 1, 1, 0] \n",
      "\n",
      "[16173634, 1, 1, 0] \n",
      "\n",
      "[16183780, 1, 1, 0] \n",
      "\n",
      "[16183790, 1, 1, 0] \n",
      "\n",
      "[16203241, 1, 1, 0] \n",
      "\n",
      "[16212160, 1, 0, 1] \n",
      "\n",
      "[16214176, 1, 1, 0] \n",
      "\n",
      "[16216190, 0, 0, 1] \n",
      "\n",
      "[16216383, 0, 1, 0] \n",
      "\n",
      "[16217414, 0, 1, 0] \n",
      "\n",
      "[16217477, 1, 1, 0] \n",
      "\n",
      "[16219080, 0, 0, 1] \n",
      "\n",
      "[16219632, 1, 0, 1] \n",
      "\n",
      "[16223527, 1, 0, 1] \n",
      "\n",
      "[16228164, 1, 1, 0] \n",
      "\n",
      "[16228270, 1, 1, 0] \n",
      "\n",
      "[16230082, 1, 1, 0] \n",
      "\n",
      "[16230525, 1, 1, 0] \n",
      "\n",
      "[16240092, 1, 1, 0] \n",
      "\n",
      "[16246137, 1, 0, 1] \n",
      "\n",
      "[16258516, 0, 1, 0] \n",
      "\n",
      "[16267436, 0, 1, 0] \n",
      "\n",
      "[16278102, 1, 1, 0] \n",
      "\n",
      "[16279410, 1, 1, 0] \n",
      "\n",
      "[16296459, 0, 0, 1] \n",
      "\n",
      "[16305058, 0, 1, 0] \n",
      "\n",
      "[16335320, 1, 1, 0] \n",
      "\n",
      "[16335338, 1, 1, 0] \n",
      "\n",
      "[16337995, 1, 1, 0] \n",
      "\n",
      "[16350349, 1, 0, 1] \n",
      "\n",
      "[16360995, 1, 0, 1] \n",
      "\n",
      "[16361304, 1, 0, 1] \n",
      "\n",
      "[16363031, 1, 0, 1] \n",
      "\n",
      "[16364324, 1, 0, 1] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "diff_list_s2=[]\n",
    "\n",
    "for block_i in range(len(allele_blocks_s2)): # \n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "\n",
    "    for var_i, varpos in enumerate(varpos_block_s2):\n",
    "        allele_s2= allele_block_s2[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "        \n",
    "            if allele_s2!=allele_s5:\n",
    "                \n",
    "                for block_i in range(number_phaseblocks_s1): # \n",
    "   \n",
    "                    allele_block_s1=allele_blocks_s1[block_i]\n",
    "                    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "                    if varpos in varpos_block_s1:\n",
    "                        allele_s1= allele_block_s1[varpos_block_s2.index(varpos)]\n",
    "                        diff_list_s2.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                        print([varpos,allele_s1,allele_s2,allele_s5],\"\\n\")\n",
    "                        break\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56 23 0 33\n"
     ]
    }
   ],
   "source": [
    "hifiok_illno=[]\n",
    "hifino_illno=[]\n",
    "hifino_illok=[]\n",
    "\n",
    "for [varpos, hifi,illu,parnt] in diff_list_s2:\n",
    "    if illu!=parnt and hifi==parnt:\n",
    "        hifiok_illno.append(varpos)\n",
    "    if illu!=parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illno.append(varpos)       \n",
    "    if illu==parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illok.append(varpos) \n",
    "    \n",
    "print(len(diff_list_s2),len(hifiok_illno),len(hifino_illok),len(hifino_illno))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56 0 22 33\n"
     ]
    }
   ],
   "source": [
    "hifiok_illno=[]\n",
    "hifino_illno=[]\n",
    "hifino_illok=[]\n",
    "\n",
    "for [varpos, hifi,illu,parnt] in diff_list:\n",
    "    \n",
    "    if illu!=parnt and hifi==parnt:\n",
    "        hifiok_illno.append(varpos)\n",
    "    if illu!=parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illno.append(varpos)       \n",
    "    if illu==parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illok.append(varpos) \n",
    "    \n",
    "print(len(diff_list_s2),len(hifiok_illno),len(hifino_illok),len(hifino_illno))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 1, 1]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_s5==allele_s2\n",
    "[allele_s1,allele_s5,allele_s2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "830 192 45 23 22\n",
      "99 66 33\n"
     ]
    }
   ],
   "source": [
    "similiar_count=0\n",
    "dissimiliar_count=0\n",
    "dissimiliar_count_s1ok=0\n",
    "dissimiliar_count_s2ok=0\n",
    "dissimiliar_count_exist5=0\n",
    "similiar_count_s1s2ok=0\n",
    "similiar_count_s1s2no=0\n",
    "similiar_count_exist5=0\n",
    "\n",
    "for block_i in range(number_phaseblocks_s2): # \n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "    intersected_idins1_s2=intersected_blocks_idins1_s2[block_i]\n",
    "    if intersected_idins1_s2 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s2)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s2)\n",
    "\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "        if allele_s1==allele_s2: \n",
    "            similiar_count+=1\n",
    "            if varpos in varposs_s5:\n",
    "                allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "                similiar_count_exist5+=1\n",
    "                if allele_s1==allele_s5 and allele_s2==allele_s5:\n",
    "                    similiar_count_s1s2ok+=1\n",
    "                if allele_s1!=allele_s5 and allele_s2!=allele_s5:\n",
    "                    similiar_count_s1s2no+=1\n",
    "        else:\n",
    "            #print(varpos)\n",
    "            dissimiliar_count+=1\n",
    "            if varpos in varposs_s5:\n",
    "                allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "                dissimiliar_count_exist5+=1\n",
    "                if allele_s1==allele_s5 and allele_s2!=allele_s5:\n",
    "                    dissimiliar_count_s1ok+=1\n",
    "                if allele_s1!=allele_s5 and allele_s2==allele_s5:\n",
    "                    dissimiliar_count_s2ok+=1\n",
    "                    \n",
    "\n",
    "            \n",
    "#     if similiar_count>1 and dissimiliar_count>1:\n",
    "#         print(similiar_count)\n",
    "\n",
    "print(similiar_count,dissimiliar_count,dissimiliar_count_exist5,dissimiliar_count_s1ok,dissimiliar_count_s2ok)\n",
    "print(similiar_count_exist5,similiar_count_s1s2ok,similiar_count_s1s2no)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(allele_block_s2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blockid  16099924\n",
      "Pos HiFi Illm parental \n",
      "\n",
      "16101684 1 1 0 \n",
      "\n",
      "16104385 1 1 0 \n",
      "\n",
      "16105717 1 0 0 \n",
      "\n",
      "16113547 0 0 1 \n",
      "\n",
      "16117006 0 1 1 \n",
      "\n",
      "16120074 1 0 0 \n",
      "\n",
      "16123812 1 0 0 \n",
      "\n",
      "16137152 1 0 0 \n",
      "\n",
      "16142002 1 1 0 \n",
      "\n",
      "16143370 1 1 0 \n",
      "\n",
      "16143995 1 1 0 \n",
      "\n",
      "16144768 1 1 0 \n",
      "\n",
      "16144952 1 1 0 \n",
      "\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "\n",
    "diff_list=[]\n",
    "count_simi=0\n",
    "block_i=1\n",
    "print(\"blockid \",blockid_blocks_s1[block_i])\n",
    "\n",
    "allele_block_s1=allele_blocks_s1[block_i]\n",
    "varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "print(\"Pos\",\"HiFi\",\"Illm\",\"parental\",\"\\n\")\n",
    "for var_i, varpos in enumerate(varpos_block_s1):\n",
    "    allele_s1= allele_block_s1[var_i]\n",
    "\n",
    "    if varpos in varposs_s5:\n",
    "        allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "\n",
    "        if allele_s1!=allele_s5:\n",
    "\n",
    "            for block_i in range(number_phaseblocks_s2): # \n",
    "\n",
    "                allele_block_s2=allele_blocks_s2[block_i]\n",
    "                varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "                if varpos in varpos_block_s2:\n",
    "                    allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "                    diff_list.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                    print(varpos,allele_s1,allele_s2,allele_s5,\"\\n\")\n",
    "                    break\n",
    "        if allele_s1==allele_s5:\n",
    "            count_simi+=1\n",
    "            \n",
    "\n",
    "\n",
    "                \n",
    "print(count_simi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1087\n"
     ]
    }
   ],
   "source": [
    "sum1=0\n",
    "for  a in varpos_blocks_s1:\n",
    "    sum1+=len(a)\n",
    "print(sum1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum2=0\n",
    "for  a in varpos_blocks_s2:\n",
    "    sum2+=len(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1083"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
