{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Num phaseblocks in sample one is  252\n",
      "Num phaseblocks in sample two is  251\n",
      "Num phaseblocks in sample three is  6771\n",
      "Num phaseblocks in sample four is  6905\n",
      "Num vars in sample five-parental is  20903\n",
      "Num all vars is  71923\n",
      "number of occurance of weird case (=two variants are in the same block, but far from each other)  535\n"
     ]
    }
   ],
   "source": [
    "\n",
    "vcf_file_address =  \"22_5smp.vcf\" #\"a/22_2k.vcf\" #  \"22_all.vcf\"  \"a/22_2k.vcf\" #  \"a/22_2k.vcf\"#   #\"22_allps_200.vcf\"  \n",
    "vcf_file = open(vcf_file_address,'r');\n",
    "\n",
    "\n",
    "blockid_blocks_s1 = []\n",
    "allele_blocks_s1 = []\n",
    "varpos_blocks_s1 = []\n",
    "\n",
    "\n",
    "blockid_blocks_s2 = []\n",
    "allele_blocks_s2 = []\n",
    "varpos_blocks_s2 = []\n",
    "\n",
    "blockid_blocks_s3 = []\n",
    "allele_blocks_s3 = []\n",
    "varpos_blocks_s3 = []\n",
    "\n",
    "blockid_blocks_s4 = []\n",
    "allele_blocks_s4 = []\n",
    "varpos_blocks_s4 = []\n",
    "\n",
    "var_blockids1_blocks_s2 = []   # for each var in s2, shows the s1 blockid\n",
    "var_blockids1_blocks_s3 = []   # for each var in s3, shows the s1 blockid\n",
    "var_blockids1_blocks_s4 = []   # for each var in s4, shows the s1 blockid\n",
    "\n",
    "\n",
    "# for parental that it's only one block per chr so it doesn't have blockid\n",
    "alleles_s5 = []\n",
    "varposs_s5 = []\n",
    "\n",
    "\n",
    "varposs_all = []\n",
    "\n",
    "num_weird_occurs=0\n",
    "\n",
    "var_lines_list = []\n",
    "\n",
    "header_lines_list = []\n",
    "for line in vcf_file:\n",
    "    line_strip = line.strip()\n",
    "    \n",
    "    if line_strip.startswith('#'):\n",
    "        header_lines_list.append(line_strip)\n",
    "        sample_names = line_strip.split('\\t')[9:]   # last line of header contains sample name\n",
    "        #print(sample_names)\n",
    "        \n",
    "    else:\n",
    "        line_parts = line_strip.split('\\t')\n",
    "        var_lines_list.append(line_parts)\n",
    "        #gt_flags, s1, sss, ss, s2, s5 = line_parts[8:]\n",
    "        gt_flags, s1, s2,s3,s4, s5 = line_parts[8:]       # hifi-whatshp illumina-whatshp hifi-hapcut illumina-hapcut true\n",
    "\n",
    "        varpos = int(line_parts[1])                 # variant position\n",
    "        \n",
    "        varposs_all.append(varpos)\n",
    "        gt_flags_split = gt_flags.split(\":\")\n",
    "        \n",
    "        if \"PS\" in gt_flags_split:\n",
    "            idx_ps = gt_flags.split(\":\").index(\"PS\")\n",
    "            idx_gt = gt_flags.split(\":\").index(\"GT\")\n",
    "            \n",
    "\n",
    "                                                 ## s5- parental - last sample in the vcf file\n",
    "            split_s5 = s5.split(\":\")\n",
    "            allele_s5 = split_s5[idx_gt]\n",
    "            blockid_s5 = split_s5[idx_ps]\n",
    "            \n",
    "            if allele_s5 == \"0|1\" or allele_s5 == \"1|0\":\n",
    "                alleles_s5.append(int(allele_s5[0]))\n",
    "                varposs_s5.append(varpos)\n",
    "\n",
    "            \n",
    "            \n",
    "                                                ###  sample 1 ###\n",
    "            split_s1 = s1.split(\":\")\n",
    "            blockid_s1 = split_s1[idx_ps]\n",
    "            allele_s1 = split_s1[idx_gt]\n",
    "            \n",
    "\n",
    "            if blockid_s1 == \".\":                                 # unphased variants for sample one\n",
    "                pass\n",
    "            \n",
    "            elif blockid_s1 in blockid_blocks_s1:\n",
    "                assert allele_s1[1] == \"|\", \"allele_s1[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                #if blockid_s1 == previous_blockid_s1: previous_blockid_s1_idx has been defined in previous variant which is the last idx\n",
    "                if blockid_s1 != previous_blockid_s1:  \n",
    "                    # weird situation happens for whatshap in which two variants are in the same block but in between there is/are other phaseblock(s)\n",
    "                    # note that the unphased variants are ignored sooner and are not considered here\n",
    "                    previous_blockid_s1_idx = blockid_blocks_s1.index(blockid_s1)\n",
    "                    #print(\"two variants are in the same block, but far from each other\",varpos)                \n",
    "                    num_weird_occurs+=1\n",
    "\n",
    "                allele_blocks_s1[previous_blockid_s1_idx].append( int(allele_s1[0]) )\n",
    "                varpos_blocks_s1[previous_blockid_s1_idx].append( varpos ) \n",
    "\n",
    "                \n",
    "            else:                                                  # A new block started, it can be the first block\n",
    "                \n",
    "                allele_blocks_s1.append( [int(allele_s1[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s1.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s1.append( blockid_s1 )             # add the blockid of the first var  to the list of blockids\n",
    "                \n",
    "                previous_blockid_s1 = blockid_s1\n",
    "                previous_blockid_s1_idx = len(blockid_blocks_s1)-1  # the index of the last block id\n",
    "                \n",
    "\n",
    "\n",
    "                \n",
    "       \n",
    "    \n",
    "                                                ###  sample 2 ###\n",
    "            split_s2 = s2.split(\":\")\n",
    "            blockid_s2 = split_s2[idx_ps]\n",
    "            allele_s2 = split_s2[idx_gt]\n",
    "            if blockid_s2 == \".\":\n",
    "                pass\n",
    "            elif blockid_s2 in blockid_blocks_s2:\n",
    "                assert allele_s2[1] == \"|\", \"allele_s2[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                if blockid_s2 != previous_blockid_s2:  \n",
    "                    previous_blockid_s2_idx = blockid_blocks_s2.index(blockid_s2)\n",
    "                    num_weird_occurs+=1\n",
    "                allele_blocks_s2[previous_blockid_s2_idx].append( int(allele_s2[0]) )\n",
    "                varpos_blocks_s2[previous_blockid_s2_idx].append( varpos ) \n",
    "                var_blockids1_blocks_s2[previous_blockid_s2_idx].append( blockid_s1 )\n",
    "                \n",
    "            else:                                                 # A new block started   (it can be the first block)\n",
    "                allele_blocks_s2.append( [int(allele_s2[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s2.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s2.append( blockid_s2 )             # add the blockid of the first var  to the list of blockids\n",
    "                var_blockids1_blocks_s2.append( [blockid_s1] )  \n",
    "                previous_blockid_s2 = blockid_s2\n",
    "                previous_blockid_s2_idx = len(blockid_blocks_s2)-1   # the index of the last block id\n",
    "                \n",
    "\n",
    "                                                ###  sample 3 ###\n",
    "            split_s3 = s3.split(\":\")\n",
    "            blockid_s3 = split_s3[idx_ps]\n",
    "            allele_s3 = split_s3[idx_gt]\n",
    "            if blockid_s3 == \".\":\n",
    "                pass\n",
    "            elif blockid_s3 in blockid_blocks_s3:\n",
    "                assert allele_s3[1] == \"|\", \"allele_s3[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                if blockid_s3 != previous_blockid_s3:  \n",
    "                    previous_blockid_s3_idx = blockid_blocks_s3.index(blockid_s3)\n",
    "                    num_weird_occurs+=1\n",
    "                allele_blocks_s3[previous_blockid_s3_idx].append( int(allele_s3[0]) )\n",
    "                varpos_blocks_s3[previous_blockid_s3_idx].append( varpos ) \n",
    "                var_blockids1_blocks_s3[previous_blockid_s3_idx].append( blockid_s1 )\n",
    "                \n",
    "            else:                                                 # A new block started   (it can be the first block)\n",
    "                allele_blocks_s3.append( [int(allele_s3[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s3.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s3.append( blockid_s3 )             # add the blockid of the first var  to the list of blockids\n",
    "                var_blockids1_blocks_s3.append( [blockid_s1] )  \n",
    "                previous_blockid_s3 = blockid_s3\n",
    "                previous_blockid_s3_idx = len(blockid_blocks_s3)-1   # the index of the last block id\n",
    "                \n",
    "   \n",
    "    \n",
    "                                                ###  sample 3 ###\n",
    "            split_s4 = s4.split(\":\")\n",
    "            blockid_s4 = split_s4[idx_ps]\n",
    "            allele_s4 = split_s4[idx_gt]\n",
    "            if blockid_s4 == \".\":\n",
    "                pass\n",
    "            elif blockid_s4 in blockid_blocks_s4:\n",
    "                assert allele_s4[1] == \"|\", \"allele_s4[1] should be | and so phased. A non-phased variant is in the phase block\"\n",
    "                \n",
    "                if blockid_s4 != previous_blockid_s4:  \n",
    "                    previous_blockid_s4_idx = blockid_blocks_s4.index(blockid_s4)\n",
    "                    num_weird_occurs+=1\n",
    "                allele_blocks_s4[previous_blockid_s4_idx].append( int(allele_s4[0]) )\n",
    "                varpos_blocks_s4[previous_blockid_s4_idx].append( varpos ) \n",
    "                var_blockids1_blocks_s4[previous_blockid_s4_idx].append( blockid_s1 )\n",
    "                \n",
    "            else:                                                 # A new block started   (it can be the first block)\n",
    "                allele_blocks_s4.append( [int(allele_s4[0])] )     # alleles of a phase block (all vars are hetrozygous) \n",
    "                varpos_blocks_s4.append( [varpos] )                # variant posistion of a block\n",
    "                blockid_blocks_s4.append( blockid_s4 )             # add the blockid of the first var  to the list of blockids\n",
    "                var_blockids1_blocks_s4.append( [blockid_s1] )  \n",
    "                previous_blockid_s4 = blockid_s4\n",
    "                previous_blockid_s4_idx = len(blockid_blocks_s4)-1   # the index of the last block id\n",
    "                \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(\"Num phaseblocks in sample one is \",len(allele_blocks_s1))\n",
    "print(\"Num phaseblocks in sample two is \",len(allele_blocks_s2))\n",
    "print(\"Num phaseblocks in sample three is \",len(allele_blocks_s3))\n",
    "print(\"Num phaseblocks in sample four is \",len(allele_blocks_s4))\n",
    "print(\"Num vars in sample five-parental is \",len(varposs_s5))\n",
    "print(\"Num all vars is \",len(varposs_all))\n",
    "\n",
    "print(\"number of occurance of weird case (=two variants are in the same block, but far from each other) \", num_weird_occurs)\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for test code\n",
    "# cd /Volumes/work/myjupyter/phaseyou\n",
    "#grep \"46014176\" 22_all.vcf\n",
    "\n",
    "#blockid_blocks_s2.index(\"46014176\")\n",
    "#print(allele_blocks_s2[5568],\"\\n\",varpos_blocks_s2[5568])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "small [31182721]\n",
      "small [36247796]\n",
      "small [47917314]\n",
      "37559\n",
      "39025\n"
     ]
    }
   ],
   "source": [
    "#%pwd\n",
    "num_phased_s1=0\n",
    "for varpos_block_s1 in varpos_blocks_s1:\n",
    "    num_phased_s1+=len(varpos_block_s1)\n",
    "    if len(varpos_block_s1)<2:\n",
    "        print(\"small\",varpos_block_s1)\n",
    "print(num_phased_s1)\n",
    "\n",
    "\n",
    "# num_phased_s2=0\n",
    "# for allele_block_s2 in allele_blocks_s2:\n",
    "#     num_phased_s2+=len(allele_block_s2)\n",
    "#     if len(allele_block_s2)<2:\n",
    "#         print(\"small,s2\",len(allele_block_s2))\n",
    "# print(num_phased_s2)\n",
    "\n",
    "num_phased_a=0\n",
    "for a in var_blockids1_blocks_s2:\n",
    "    num_phased_a+=len(a)\n",
    "print(num_phased_a)   \n",
    "#len(var_blockids1_blocks_s2[100])\n",
    "#print(var_blockids1_blocks_s2[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_blocks_idins1_s2=[] # for each block in s2, the corresponding s1 block id\n",
    "\n",
    "for var_blockids1_blocks_s2_i in var_blockids1_blocks_s2:\n",
    "    var_blockids1_blocks_s2_i_set=set(var_blockids1_blocks_s2_i)\n",
    "    \n",
    "    if '.' in var_blockids1_blocks_s2_i_set:\n",
    "        var_blockids1_blocks_s2_i_set.discard('.')\n",
    "        \n",
    "    if len(var_blockids1_blocks_s2_i_set)==1:\n",
    "        intersected_idins1_s2=list(var_blockids1_blocks_s2_i_set)[0]\n",
    "    if len(var_blockids1_blocks_s2_i_set)==0:\n",
    "        intersected_idins1_s2=0\n",
    "    if len(var_blockids1_blocks_s2_i_set)>1:\n",
    "        # find the most frequent blockid in the phaseblock\n",
    "        intersected_idins1_s2=max(var_blockids1_blocks_s2_i_set, key=var_blockids1_blocks_s2_i.count) \n",
    "    \n",
    "        \n",
    "    intersected_blocks_idins1_s2.append(intersected_idins1_s2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_blocks_idins1_s3=[] # for each block in s3, the corresponding s1 block id\n",
    "\n",
    "for var_blockids1_blocks_s3_i in var_blockids1_blocks_s3:\n",
    "    var_blockids1_blocks_s3_i_set=set(var_blockids1_blocks_s3_i)\n",
    "    \n",
    "    if '.' in var_blockids1_blocks_s3_i_set:\n",
    "        var_blockids1_blocks_s3_i_set.discard('.')\n",
    "        \n",
    "    if len(var_blockids1_blocks_s3_i_set)==1:\n",
    "        intersected_idins1_s3=list(var_blockids1_blocks_s3_i_set)[0]\n",
    "    if len(var_blockids1_blocks_s3_i_set)==0:\n",
    "        intersected_idins1_s3=0\n",
    "    if len(var_blockids1_blocks_s3_i_set)>1:\n",
    "        # find the most frequent blockid in the phaseblock\n",
    "        intersected_idins1_s3=max(var_blockids1_blocks_s3_i_set, key=var_blockids1_blocks_s3_i.count) \n",
    "    \n",
    "        \n",
    "    intersected_blocks_idins1_s3.append(intersected_idins1_s3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "intersected_blocks_idins1_s4=[] # for each block in s4, the corresponding s1 block id\n",
    "\n",
    "for var_blockids1_blocks_s4_i in var_blockids1_blocks_s4:\n",
    "    var_blockids1_blocks_s4_i_set=set(var_blockids1_blocks_s4_i)\n",
    "    \n",
    "    if '.' in var_blockids1_blocks_s4_i_set:\n",
    "        var_blockids1_blocks_s4_i_set.discard('.')\n",
    "        \n",
    "    if len(var_blockids1_blocks_s4_i_set)==1:\n",
    "        intersected_idins1_s4=list(var_blockids1_blocks_s4_i_set)[0]\n",
    "    if len(var_blockids1_blocks_s4_i_set)==0:\n",
    "        intersected_idins1_s4=0\n",
    "    if len(var_blockids1_blocks_s4_i_set)>1:\n",
    "        # find the most frequent blockid in the phaseblock\n",
    "        intersected_idins1_s4=max(var_blockids1_blocks_s4_i_set, key=var_blockids1_blocks_s4_i.count) \n",
    "    \n",
    "        \n",
    "    intersected_blocks_idins1_s4.append(intersected_idins1_s4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(blockid_blocks_s1)\n",
    "#for i in range(len(allele_blocks_s1)):\n",
    "#    print(len(allele_blocks_s1[i]))\n",
    "    \n",
    "#print(blockid_blocks_s2)\n",
    "#for i in range(len(allele_blocks_s2)):\n",
    "#    print(len(allele_blocks_s2[i]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# re-ordering  the phase block allel based on parent\n",
    "\n",
    "## important ## \n",
    "# if we don't have the parental , how we can reorder\n",
    "\n",
    "\n",
    "number_phaseblocks_s1= len(allele_blocks_s1)\n",
    "\n",
    "#allele_blocks_s1_new=list(allele_blocks_s1)\n",
    "for block_i in range(number_phaseblocks_s1): # \n",
    "   \n",
    "    allele_block_s1=allele_blocks_s1[block_i]\n",
    "    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "\n",
    "    similiar_count=0\n",
    "    diss_count=0\n",
    "    for var_i, varpos in enumerate(varpos_block_s1):\n",
    "        allele_s1= allele_block_s1[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "            if allele_s1 == allele_s5:\n",
    "                similiar_count+=1\n",
    "            else: \n",
    "                diss_count+=1\n",
    "                \n",
    "    #print(similiar_count, diss_count)          \n",
    "    if similiar_count < diss_count:\n",
    "        allele_block_s1_reordered = [1-i for i in allele_block_s1] # converting 0->1 and 1->0\n",
    "        allele_blocks_s1[block_i]=allele_block_s1_reordered\n",
    "        #print(\"it needs to re-order\")\n",
    "    \n",
    "#         allele_block_s1_new = [1-i for i in allele_block_s1]\n",
    "#         allele_blocks_s1_new[block_i]=allele_block_s1_new\n",
    "        \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_phaseblocks_s2= len(allele_blocks_s2)\n",
    "\n",
    "for block_i in range(number_phaseblocks_s2): # \n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "    intersected_idins1_s2=intersected_blocks_idins1_s2[block_i]\n",
    "    if intersected_idins1_s2 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s2)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s2)\n",
    "    similiar_count=0\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "        if allele_s1==allele_s2: similiar_count+=1\n",
    "    #print(similiar_count/len(intersect_varpos))\n",
    "    if similiar_count < len(intersect_varpos)/2:\n",
    "        #print(\"need\", similiar_count)\n",
    "        allele_block_s2_reordered= [1-i for i in allele_block_s2]\n",
    "        allele_blocks_s2[block_i]=allele_block_s2_reordered\n",
    "        #print(\"it needs to re-order\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_phaseblocks_s3= len(allele_blocks_s3)\n",
    "\n",
    "for block_i in range(number_phaseblocks_s3): # \n",
    "   \n",
    "    allele_block_s3=allele_blocks_s3[block_i]\n",
    "    varpos_block_s3=varpos_blocks_s3[block_i]\n",
    "\n",
    "    intersected_idins1_s3=intersected_blocks_idins1_s3[block_i]\n",
    "    if intersected_idins1_s3 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s3)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s3)\n",
    "    similiar_count=0\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s3= allele_block_s3[varpos_block_s3.index(varpos)]\n",
    "        if allele_s1==allele_s3: similiar_count+=1\n",
    "    #print(similiar_count/len(intersect_varpos))\n",
    "    if similiar_count < len(intersect_varpos)/2:\n",
    "        #print(\"need\", similiar_count)\n",
    "        allele_block_s3_reordered= [1-i for i in allele_block_s3]\n",
    "        allele_blocks_s3[block_i]=allele_block_s3_reordered\n",
    "        #print(\"it needs to re-order\")\n",
    "\n",
    "\n",
    "        \n",
    "number_phaseblocks_s4= len(allele_blocks_s4)\n",
    "\n",
    "for block_i in range(number_phaseblocks_s4): # \n",
    "   \n",
    "    allele_block_s4=allele_blocks_s4[block_i]\n",
    "    varpos_block_s4=varpos_blocks_s4[block_i]\n",
    "\n",
    "    intersected_idins1_s4=intersected_blocks_idins1_s4[block_i]\n",
    "    if intersected_idins1_s4 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s4)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s4)\n",
    "    similiar_count=0\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s4= allele_block_s4[varpos_block_s4.index(varpos)]\n",
    "        if allele_s1==allele_s4: similiar_count+=1\n",
    "    #print(similiar_count/len(intersect_varpos))\n",
    "    if similiar_count < len(intersect_varpos)/2:\n",
    "        #print(\"need\", similiar_count)\n",
    "        allele_block_s4_reordered= [1-i for i in allele_block_s4]\n",
    "        allele_blocks_s4[block_i]=allele_block_s4_reordered\n",
    "        #print(\"it needs to re-order\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix_row_lists=[]\n",
    "for varpos in varposs_all:\n",
    "    # for increasing the speed we can use the information that the phase blocks are in order.\n",
    "    # but not for those weird phase block\n",
    "\n",
    "\n",
    "    allele_s1='.'\n",
    "    blockid_s1='.'\n",
    "    for block_i_s1 in range(len(blockid_blocks_s1)):\n",
    "        allele_block_s1 = allele_blocks_s1[block_i_s1]\n",
    "        varpos_block_s1 = varpos_blocks_s1[block_i_s1]\n",
    "\n",
    "        if varpos in varpos_block_s1:\n",
    "            allele_s1 = allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "            blockid_s1 = blockid_blocks_s1[block_i_s1]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "            \n",
    "    allele_s2='.'\n",
    "    blockid_s2='.'\n",
    "    for block_i_s2 in range(len(blockid_blocks_s2)):\n",
    "        allele_block_s2 = allele_blocks_s2[block_i_s2]\n",
    "        varpos_block_s2 = varpos_blocks_s2[block_i_s2]\n",
    "\n",
    "        if varpos in varpos_block_s2:\n",
    "            allele_s2 = allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "            blockid_s2 = blockid_blocks_s2[block_i_s2]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "    allele_s3='.'\n",
    "    blockid_s3='.'\n",
    "    for block_i_s3 in range(len(blockid_blocks_s3)):\n",
    "        allele_block_s3 = allele_blocks_s3[block_i_s3]\n",
    "        varpos_block_s3 = varpos_blocks_s3[block_i_s3]\n",
    "\n",
    "        if varpos in varpos_block_s3:\n",
    "            allele_s3 = allele_block_s3[varpos_block_s3.index(varpos)]\n",
    "            blockid_s3 = blockid_blocks_s3[block_i_s3]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "    allele_s4='.'\n",
    "    blockid_s4='.'\n",
    "    for block_i_s4 in range(len(blockid_blocks_s4)):\n",
    "        allele_block_s4 = allele_blocks_s4[block_i_s4]\n",
    "        varpos_block_s4 = varpos_blocks_s4[block_i_s4]\n",
    "\n",
    "        if varpos in varpos_block_s4:\n",
    "            allele_s4 = allele_block_s4[varpos_block_s4.index(varpos)]\n",
    "            blockid_s4 = blockid_blocks_s4[block_i_s4]\n",
    "            break # when var is found, no need to continue for loop\n",
    "\n",
    "    \n",
    "            \n",
    "    if varpos in varposs_s5:\n",
    "        allele_s5 = alleles_s5[varposs_s5.index(varpos)]\n",
    "    else: allele_s5= '.'\n",
    "\n",
    "    if not (allele_s1=='.' and allele_s2=='.' and allele_s3=='.' and allele_s4=='.' and allele_s5=='.'): \n",
    "        matrix_row_list=[str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "                         str(allele_s2)+\":\"+str(blockid_s2),str(allele_s3)+\":\"+str(blockid_s3),\n",
    "                         str(allele_s4)+\":\"+str(blockid_s4), str(allele_s5)] \n",
    "        matrix_row_lists.append(matrix_row_list)\n",
    "\n",
    "        \n",
    "# one option can be for loop over blockid_blocks_s1 instead of varposs_all, then search in  blockid_blocks_s1 also for printing sort them to overcome weird phase block\n",
    "# the issue will be for those variant that are not phased in blockid_blocks_s1 but in blockid_blocks_s2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16102810 False\n",
      "16103146 False\n",
      "16103299 False\n",
      "16115554 False\n",
      "16117006 False\n",
      "16203542 False\n",
      "16345401 False\n",
      "16347827 False\n",
      "16351604 False\n",
      "16442657 False\n",
      "16490618 True\n",
      "16499684 False\n",
      "16499788 False\n",
      "16506873 True\n",
      "16512581 False\n",
      "16528995 True\n",
      "16553221 True\n",
      "16558766 True\n",
      "16561186 True\n",
      "16564036 True\n",
      "16590588 False\n",
      "16623522 False\n",
      "16664149 True\n",
      "16864037 False\n",
      "16864095 False\n",
      "16864325 False\n",
      "16864351 False\n",
      "16864741 False\n",
      "16882859 True\n",
      "16883072 True\n",
      "16883221 True\n",
      "16884830 True\n",
      "17392568 False\n",
      "17614938 False\n",
      "17617072 False\n",
      "17627489 False\n",
      "17760350 False\n",
      "17760620 False\n",
      "17893503 False\n",
      "18309097 False\n",
      "18359712 False\n",
      "18532438 False\n",
      "18605939 False\n",
      "18764533 True\n",
      "18764558 False\n",
      "18764739 False\n",
      "18770724 True\n",
      "18807561 False\n",
      "18807958 False\n",
      "18830799 False\n",
      "18830821 False\n",
      "18831309 False\n",
      "18831620 False\n",
      "18832024 False\n",
      "18832515 False\n",
      "18837115 False\n",
      "18837431 False\n",
      "18837489 False\n",
      "18838215 False\n",
      "18838256 False\n",
      "18839101 False\n",
      "18840046 False\n",
      "18840051 False\n",
      "18840378 False\n",
      "18840440 False\n",
      "18841317 False\n",
      "18841329 False\n",
      "18841614 False\n",
      "18841624 False\n",
      "18841654 True\n",
      "18841811 False\n",
      "18841874 False\n",
      "18841900 True\n",
      "18841931 False\n",
      "18842061 False\n",
      "18842353 False\n",
      "18842453 False\n",
      "18845285 False\n",
      "18845808 False\n",
      "18848106 False\n",
      "18848136 False\n",
      "18848286 False\n",
      "18860847 False\n",
      "18860859 False\n",
      "18860901 False\n",
      "18860994 False\n",
      "18863383 False\n",
      "18935180 False\n",
      "18935320 False\n",
      "18935759 True\n",
      "18935766 False\n",
      "18935842 True\n",
      "19014037 True\n",
      "19072315 False\n",
      "19558198 False\n",
      "19640218 False\n",
      "19713300 False\n",
      "19713760 False\n",
      "19713861 False\n",
      "19714874 False\n",
      "19895427 False\n",
      "19907099 False\n",
      "19907118 False\n",
      "19907192 False\n",
      "20063076 False\n",
      "20150717 False\n",
      "20150768 False\n",
      "20171881 False\n",
      "20171961 False\n",
      "20171974 False\n",
      "20197335 False\n",
      "20203378 False\n",
      "20234016 False\n",
      "20292393 False\n",
      "20296346 False\n",
      "20296671 False\n",
      "20324038 False\n",
      "20324294 False\n",
      "20324425 False\n",
      "20324440 False\n",
      "20325425 False\n",
      "20325434 False\n",
      "20655662 False\n",
      "20656969 False\n",
      "20657880 False\n",
      "20658429 False\n",
      "20658480 False\n",
      "20658600 False\n",
      "20663893 False\n",
      "20664596 False\n",
      "20665108 False\n",
      "20765333 False\n",
      "21468861 True\n",
      "21528946 False\n",
      "21535771 False\n",
      "21999108 False\n",
      "22038765 False\n",
      "22478713 False\n",
      "22537472 False\n",
      "22621673 False\n",
      "23055331 False\n",
      "23659001 False\n",
      "23659043 False\n",
      "23667463 False\n",
      "23667474 False\n",
      "23791099 False\n",
      "24009108 False\n",
      "24009160 False\n",
      "24109136 False\n",
      "25000163 False\n",
      "25001173 False\n",
      "25001185 False\n",
      "25015744 False\n",
      "25043308 False\n",
      "25043518 False\n",
      "25043618 False\n",
      "25043636 False\n",
      "blockIds of s1 and s2 don't match:  25066913 . .\n",
      "blockIds of s1 and s2 don't match:  25067393 . .\n",
      "blockIds of s1 and s2 don't match:  25067451 . .\n",
      "25310044 False\n",
      "25614672 False\n",
      "27415439 False\n",
      "28021693 False\n",
      "28065092 False\n",
      "28066707 False\n",
      "28066714 False\n",
      "30339761 False\n",
      "30339774 False\n",
      "31019844 False\n",
      "31742635 False\n",
      "31742823 False\n",
      "32858178 False\n",
      "32899899 False\n",
      "32961097 False\n",
      "34974844 False\n",
      "35227247 False\n",
      "35228637 False\n",
      "35859460 False\n",
      "35887883 False\n",
      "35887884 False\n",
      "35974403 False\n",
      "36007921 False\n",
      "blockIds of s1 and s2 don't match:  36490708 . .\n",
      "36522615 False\n",
      "blockIds of s1 and s2 don't match:  36541000 . .\n",
      "blockIds of s1 and s2 don't match:  36615084 . .\n",
      "blockIds of s1 and s2 don't match:  36617974 . .\n",
      "blockIds of s1 and s2 don't match:  36624461 . .\n",
      "blockIds of s1 and s2 don't match:  36625910 . .\n",
      "blockIds of s1 and s2 don't match:  36954192 . .\n",
      "37269022 False\n",
      "37269033 False\n",
      "37269106 False\n",
      "37355157 False\n",
      "37355201 False\n",
      "37355215 False\n",
      "blockIds of s1 and s2 don't match:  37409576 . .\n",
      "blockIds of s1 and s2 don't match:  37425201 . .\n",
      "37925124 False\n",
      "37962312 False\n",
      "38047333 False\n",
      "38506074 False\n",
      "39072681 True\n",
      "blockIds of s1 and s2 don't match:  39323324 . .\n",
      "39376269 False\n",
      "39998023 False\n",
      "40409112 False\n",
      "blockIds of s1 and s2 don't match:  40485477 . .\n",
      "40932398 False\n",
      "40932566 False\n",
      "42027589 False\n",
      "42859935 False\n",
      "42871997 False\n",
      "42872003 False\n",
      "blockIds of s1 and s2 don't match:  43163714 . .\n",
      "blockIds of s1 and s2 don't match:  43256165 . .\n",
      "blockIds of s1 and s2 don't match:  43270877 . .\n",
      "blockIds of s1 and s2 don't match:  43295268 . .\n",
      "blockIds of s1 and s2 don't match:  43323582 . .\n",
      "blockIds of s1 and s2 don't match:  43327579 . .\n",
      "blockIds of s1 and s2 don't match:  43340279 . .\n",
      "blockIds of s1 and s2 don't match:  43375700 . .\n",
      "43379343 False\n",
      "43475081 False\n",
      "43498992 False\n",
      "blockIds of s1 and s2 don't match:  43500440 . .\n",
      "blockIds of s1 and s2 don't match:  43504326 . .\n",
      "blockIds of s1 and s2 don't match:  43583598 . .\n",
      "blockIds of s1 and s2 don't match:  43629518 . .\n",
      "blockIds of s1 and s2 don't match:  43792249 . .\n",
      "43808328 False\n",
      "43887487 False\n",
      "43887497 False\n",
      "43887518 False\n",
      "43891373 False\n",
      "43891419 False\n",
      "43891434 False\n",
      "43891452 False\n",
      "43892052 False\n",
      "43892075 False\n",
      "43892090 False\n",
      "43892103 False\n",
      "43892144 False\n",
      "43892157 False\n",
      "43892208 False\n",
      "43892284 False\n",
      "43892311 False\n",
      "43892330 False\n",
      "43892364 False\n",
      "43892472 False\n",
      "43892630 False\n",
      "43892744 False\n",
      "43892849 False\n",
      "43895454 False\n",
      "43895689 False\n",
      "43895707 False\n",
      "43895713 False\n",
      "43895767 False\n",
      "43895897 False\n",
      "43895944 False\n",
      "43896049 False\n",
      "43896081 False\n",
      "43896100 False\n",
      "blockIds of s1 and s2 don't match:  43896682 . .\n",
      "blockIds of s1 and s2 don't match:  43899415 . .\n",
      "blockIds of s1 and s2 don't match:  43938882 . .\n",
      "44126108 False\n",
      "44126306 False\n",
      "44201623 False\n",
      "44231005 False\n",
      "44279987 False\n",
      "blockIds of s1 and s2 don't match:  44290480 . .\n",
      "blockIds of s1 and s2 don't match:  44324727 . .\n",
      "blockIds of s1 and s2 don't match:  44360007 . .\n",
      "blockIds of s1 and s2 don't match:  44378670 . .\n",
      "44702691 False\n",
      "44702794 False\n",
      "45199559 False\n",
      "45512149 False\n",
      "45624885 False\n",
      "45698610 False\n",
      "46003447 False\n",
      "46003699 False\n",
      "46963052 False\n",
      "46963107 False\n",
      "46963260 False\n",
      "47002482 False\n",
      "47010953 False\n",
      "47012876 False\n",
      "47012911 False\n",
      "47012989 False\n",
      "47013007 False\n",
      "47013023 False\n",
      "47459365 False\n",
      "47775088 False\n",
      "47801249 False\n",
      "47853172 False\n",
      "48051797 False\n",
      "48188027 False\n",
      "48600260 False\n",
      "48600367 True\n",
      "48978536 False\n",
      "blockIds of s1 and s2 don't match:  48995974 . .\n",
      "49132160 False\n",
      "49475490 False\n",
      "49514692 False\n",
      "49922352 False\n",
      "49922744 False\n",
      "49941350 False\n",
      "49941793 False\n",
      "50467415 False\n",
      "50467481 False\n",
      "50467567 False\n",
      "50501957 False\n",
      "50927104 False\n",
      "50927155 False\n",
      "50977233 False\n",
      "51001102 False\n",
      "51001118 False\n",
      "51082446 False\n",
      "51082532 False\n",
      "51082554 False\n",
      "51082563 False\n",
      "51082597 False\n",
      "51082606 False\n",
      "51082614 False\n",
      "51082649 False\n",
      "51082692 False\n",
      "51125106 False\n",
      "51146139 True\n",
      "51225514 False\n",
      "51242557 False\n"
     ]
    }
   ],
   "source": [
    "result_consensus =[]\n",
    "\n",
    "changes_varpos_pre=[]\n",
    "changes_varpos_next=[]\n",
    "\n",
    "matrix_row_lists_prd =[]\n",
    "for row_i, matrix_row_list in enumerate(matrix_row_lists):\n",
    "    varpos, s1, s2, s3, s4, allele_s5 = matrix_row_list\n",
    "    \n",
    "    allele_s1='.';blockid_s1='.'\n",
    "    allele_s2='.';blockid_s2='.'\n",
    "    allele_s3='.';blockid_s3='.'\n",
    "    allele_s4='.';blockid_s4='.'\n",
    "    if s1 != '.:.':\n",
    "        allele_s1 = s1[0]\n",
    "        blockid_s1 = s1[2:]\n",
    "        blockid_s1_last_known = blockid_s1\n",
    "        \n",
    "    if s2 != '.:.':\n",
    "        allele_s2 = s2[0]\n",
    "        blockid_s2 = s2[2:]        \n",
    "    if s3 != '.:.':\n",
    "        allele_s3 = s3[0]\n",
    "        blockid_s3 = s3[2:]\n",
    "    if s4 != '.:.':\n",
    "        allele_s4 = s4[0]\n",
    "        blockid_s4 = s4[2:]\n",
    "    \n",
    "    allele_s_p = allele_s1\n",
    "    blockid_s_p = blockid_s1\n",
    "    \n",
    "    if allele_s1 == '.' and allele_s2 != \".\":\n",
    "\n",
    "        if blockid_s2 == blockid_s1_last_known:\n",
    "            allele_s_p = allele_s2\n",
    "            blockid_s_p = blockid_s2\n",
    "            changes_varpos_pre.append(varpos)\n",
    "        else:             # check next variant, \n",
    "\n",
    "            blockid_s1_next = matrix_row_lists[row_i+1][1][2:]\n",
    "            blockid_s2_next = matrix_row_lists[row_i+1][2][2:]\n",
    "            if blockid_s2_next == blockid_s2:\n",
    "                allele_s_p = allele_s2\n",
    "                blockid_s_p = blockid_s2 \n",
    "                changes_varpos_next.append(varpos)\n",
    "\n",
    "            else:\n",
    "                print(\"blockIds of s1 and s2 don't match: \",varpos,allele_s_p,blockid_s_p,)\n",
    "            \n",
    "    if allele_s1 != '.' :\n",
    "        if allele_s1 != allele_s2 and allele_s2 == allele_s3 and allele_s3 == allele_s4:\n",
    "            result_consensus.append(allele_s1 == allele_s5)\n",
    "            print(varpos, allele_s1 == allele_s5)\n",
    "            allele_s_p = str(1-int(allele_s_p))\n",
    "\n",
    "   \n",
    "            \n",
    "            \n",
    "        \n",
    "    matrix_row_list_prd = [str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "                                  str(allele_s2)+\":\"+str(blockid_s2),str(allele_s3)+\":\"+str(blockid_s3),\n",
    "                                  str(allele_s4)+\":\"+str(blockid_s4), str(allele_s5),\n",
    "                                  str(allele_s_p)+\":\"+str(blockid_s_p)] \n",
    "    \n",
    "    matrix_row_lists_prd.append(matrix_row_list_prd)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23, 298)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(result_consensus), len(result_consensus)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "252 0\n"
     ]
    }
   ],
   "source": [
    "print(len(changes_varpos_pre),len(changes_varpos_next))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "changes_varpos_next[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "matrix_address=vcf_file_address+\"_matrix5_p2.txt\"\n",
    "\n",
    "matrix_file=open(matrix_address,'w');\n",
    "matrix_file.write(\"\\t\\t\".join([\"Pos\",\"HiFiWsp\",\"HiFiHpct\",\"IllmWsp\",\"IllmHpct\",\"parental\",\"predicted\"])+\"\\n\") # \"Pos\",\"ONT\",\"CLR\",\"HiFi\",\"Illm\",\"parental\"\n",
    "for matrix_row_list in matrix_row_lists_prd:\n",
    "\n",
    "        matrix_file.write(\"\\t\".join(matrix_row_list)+\"\\n\")\n",
    "\n",
    "matrix_file.close()\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'22_5smp.vcf'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file_address"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blockIds of s1 and s2 don't match:  25066913 . .\n",
      "blockIds of s1 and s2 don't match:  25067393 . .\n",
      "blockIds of s1 and s2 don't match:  25067451 . .\n",
      "blockIds of s1 and s2 don't match:  36490708 . .\n",
      "blockIds of s1 and s2 don't match:  36541000 . .\n",
      "blockIds of s1 and s2 don't match:  36615084 . .\n",
      "blockIds of s1 and s2 don't match:  36617974 . .\n",
      "blockIds of s1 and s2 don't match:  36624461 . .\n",
      "blockIds of s1 and s2 don't match:  36625910 . .\n",
      "blockIds of s1 and s2 don't match:  36954192 . .\n",
      "blockIds of s1 and s2 don't match:  37409576 . .\n",
      "blockIds of s1 and s2 don't match:  37425201 . .\n",
      "blockIds of s1 and s2 don't match:  39323324 . .\n",
      "blockIds of s1 and s2 don't match:  40485477 . .\n",
      "blockIds of s1 and s2 don't match:  43163714 . .\n",
      "blockIds of s1 and s2 don't match:  43256165 . .\n",
      "blockIds of s1 and s2 don't match:  43270877 . .\n",
      "blockIds of s1 and s2 don't match:  43295268 . .\n",
      "blockIds of s1 and s2 don't match:  43323582 . .\n",
      "blockIds of s1 and s2 don't match:  43327579 . .\n",
      "blockIds of s1 and s2 don't match:  43340279 . .\n",
      "blockIds of s1 and s2 don't match:  43375700 . .\n",
      "blockIds of s1 and s2 don't match:  43500440 . .\n",
      "blockIds of s1 and s2 don't match:  43504326 . .\n",
      "blockIds of s1 and s2 don't match:  43583598 . .\n",
      "blockIds of s1 and s2 don't match:  43629518 . .\n",
      "blockIds of s1 and s2 don't match:  43792249 . .\n",
      "blockIds of s1 and s2 don't match:  43896682 . .\n",
      "blockIds of s1 and s2 don't match:  43899415 . .\n",
      "blockIds of s1 and s2 don't match:  43938882 . .\n",
      "blockIds of s1 and s2 don't match:  44290480 . .\n",
      "blockIds of s1 and s2 don't match:  44324727 . .\n",
      "blockIds of s1 and s2 don't match:  44360007 . .\n",
      "blockIds of s1 and s2 don't match:  44378670 . .\n",
      "blockIds of s1 and s2 don't match:  48995974 . .\n",
      "s5 2452\n",
      "s2!=s3=s4=s5: 21 s2=s3=s4=s5: 9\n",
      "s2=s3=s4!=s5: 2330 s2!=s3=s4!s5: 36\n"
     ]
    }
   ],
   "source": [
    "\n",
    "result_consensus =[]\n",
    "\n",
    "changes_varpos_pre=[]\n",
    "changes_varpos_next=[]\n",
    "\n",
    "matrix_row_lists_prd =[]\n",
    "\n",
    "\n",
    "cntr1=cntr2=cntr3=cntr4=cntr0=0\n",
    "for row_i, matrix_row_list in enumerate(matrix_row_lists):\n",
    "    varpos, s1, s2, s3, s4, allele_s5 = matrix_row_list\n",
    "    \n",
    "    allele_s1='.';blockid_s1='.'\n",
    "    allele_s2='.';blockid_s2='.'\n",
    "    allele_s3='.';blockid_s3='.'\n",
    "    allele_s4='.';blockid_s4='.'\n",
    "    if s1 != '.:.':\n",
    "        allele_s1 = s1[0]\n",
    "        blockid_s1 = s1[2:]\n",
    "        blockid_s1_last_known = blockid_s1\n",
    "        \n",
    "    if s2 != '.:.':\n",
    "        allele_s2 = s2[0]\n",
    "        blockid_s2 = s2[2:]        \n",
    "    if s3 != '.:.':\n",
    "        allele_s3 = s3[0]\n",
    "        blockid_s3 = s3[2:]\n",
    "    if s4 != '.:.':\n",
    "        allele_s4 = s4[0]\n",
    "        blockid_s4 = s4[2:]\n",
    "    \n",
    "    allele_s_p = allele_s1\n",
    "    blockid_s_p = blockid_s1\n",
    "    \n",
    "    if allele_s1 == '.' and allele_s2 != \".\":\n",
    "        \n",
    "        \n",
    "\n",
    "        if blockid_s2 == blockid_s1_last_known:\n",
    "            allele_s_p = allele_s2\n",
    "            blockid_s_p = blockid_s2\n",
    "            changes_varpos_pre.append(varpos)\n",
    "        else:             # check next variant, \n",
    "\n",
    "            blockid_s1_next = matrix_row_lists[row_i+1][1][2:]\n",
    "            blockid_s2_next = matrix_row_lists[row_i+1][2][2:]\n",
    "            if blockid_s2_next == blockid_s2:\n",
    "                allele_s_p = allele_s2\n",
    "                blockid_s_p = blockid_s2 \n",
    "                changes_varpos_next.append(varpos)\n",
    "\n",
    "            else:\n",
    "                print(\"blockIds of s1 and s2 don't match: \",varpos,allele_s_p,blockid_s_p,)\n",
    "            \n",
    "    if allele_s1 != '.' :\n",
    "        if allele_s1 != allele_s2 and allele_s2 == allele_s3 and allele_s3 == allele_s4:\n",
    "            result_consensus.append(allele_s1 == allele_s5)\n",
    "            #print(varpos, allele_s1 == allele_s5)\n",
    "            allele_s_p = str(1-int(allele_s_p))\n",
    "        \n",
    "    if allele_s1 == '.' and allele_s5 != \".\":\n",
    "    #if allele_s5 != \".\":\n",
    "        \n",
    "        \n",
    "        cntr0+=1\n",
    "        if allele_s2 != allele_s3 and allele_s3 == allele_s4 and allele_s4 == allele_s5:\n",
    "            cntr1+=1\n",
    "        if allele_s2 == allele_s3 and allele_s3 == allele_s4 and allele_s4 == allele_s5:\n",
    "            cntr2+=1\n",
    "        if allele_s2 == allele_s3 and allele_s3 == allele_s4 and allele_s4 != allele_s5:\n",
    "            cntr3+=1\n",
    "        if allele_s2 != allele_s3 and allele_s3 == allele_s4 and allele_s4 != allele_s5:\n",
    "            cntr4+=1\n",
    "        \n",
    "        \n",
    "print(\"s5\", cntr0)         \n",
    "print(\"s2!=s3=s4=s5:\",cntr1,\"s2=s3=s4=s5:\",cntr2)\n",
    "print(\"s2=s3=s4!=s5:\",cntr3,\"s2!=s3=s4!s5:\",cntr4) \n",
    "        \n",
    "#     matrix_row_list_prd = [str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "#                                   str(allele_s2)+\":\"+str(blockid_s2),str(allele_s3)+\":\"+str(blockid_s3),\n",
    "#                                   str(allele_s4)+\":\"+str(blockid_s4), str(allele_s5),\n",
    "#                                   str(allele_s_p)+\":\"+str(blockid_s_p)] \n",
    "    \n",
    "#     matrix_row_lists_prd.append(matrix_row_list_prd)\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "42002"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matrix_row_lists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71923 42002\n"
     ]
    }
   ],
   "source": [
    "varpos_list_inmatrix=[]\n",
    "for matrix_row_list in matrix_row_lists_prd:\n",
    "    varpos_list_inmatrix.append(matrix_row_list[0])\n",
    "print(len(varposs_all),len(varpos_list_inmatrix))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['22',\n",
       " '16050159',\n",
       " '.',\n",
       " 'C',\n",
       " 'T',\n",
       " '1178.67',\n",
       " '.',\n",
       " 'DPB=104;EPPR=6.51267;GTI=0;MQMR=45.7419;NS=1;NUMALT=1;ODDS=271.398;PAIREDR=0.983871;PQR=0;PRO=0;QR=2231;RO=62;RPPR=4.27115;SRF=60;SRP=120.83;SRR=2;DP=416;AB=0.403846;ABP=11.3621;AF=0.5;AO=42;CIGAR=1X;DPRA=0;EPP=13.1438;LEN=1;MEANALT=1;MQM=43.7619;PAIRED=1;PAO=0;PQA=0;QA=1554;RPL=14;RPP=13.1438;RPR=28;RUN=1;SAF=42;SAP=94.2121;SAR=0;TYPE=snp;technology.illumina=1;AN=8;AC=4',\n",
       " 'GT:DP:RO:QR:AO:QA:GL:PS:PQ:PD',\n",
       " '0|1:104:62:2231:42:1554:-138.29,0,-199.21:16050159:.:.',\n",
       " '0|1:104:.:.:.:.:.:16050159:16:1',\n",
       " '0|1:104:62:2231:42:1554:-138.29,0,-199.21:16050159:.:.',\n",
       " '1|0:104:.:.:.:.:.:16050159:100:35',\n",
       " './.:.:.:.:.:.:.:.:.:.']"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "var_lines_list_imp = []\n",
    "for var_line_list in var_lines_list: # \n",
    "    var_line_list_part=var_line_list[:9]\n",
    "    \n",
    "        \n",
    "#     matrix_row_list_prd = [str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "#                                   str(allele_s2)+\":\"+str(blockid_s2),str(allele_s3)+\":\"+str(blockid_s3),\n",
    "#                                   str(allele_s4)+\":\"+str(blockid_s4), str(allele_s5),\n",
    "#                                   str(allele_s_p)+\":\"+str(blockid_s_p)] \n",
    "    varpos = var_line_list_part[1]\n",
    "    if varpos in varpos_list_inmatrix:\n",
    "        \n",
    "        gt_flags = var_line_list[8]\n",
    "        idx_ps = gt_flags.split(\":\").index(\"PS\")\n",
    "        idx_gt = gt_flags.split(\":\").index(\"GT\")        \n",
    "        s1 = var_line_list[9] # the rest of var info is got from the s1\n",
    "        split_s1 = s1.split(\":\")\n",
    "        \n",
    "        var_idx = varpos_list_inmatrix.index(varpos)\n",
    "        matrix_row_list_prd = matrix_row_lists_prd[var_idx]\n",
    "\n",
    "        allele_s_p = matrix_row_list_prd[6][0]\n",
    "        blockid_s_p = matrix_row_list_prd[6][2:]\n",
    "        if allele_s_p != \".\":\n",
    "            split_s1[idx_gt] = allele_s_p+\"|\"+str(1-int(allele_s_p))\n",
    "            split_s1[idx_ps] = blockid_s_p \n",
    "        else:\n",
    "            split_s1[idx_gt]=\"./.\"\n",
    "            split_s1[idx_ps]=\".\"\n",
    "            \n",
    "        var_line_list_imp = split_s1\n",
    "        var_line_list_part.append(\":\".join(var_line_list_imp))\n",
    "        allele_s5 = matrix_row_list_prd[5]\n",
    "\n",
    "        if allele_s5 != \".\":\n",
    "        \n",
    "            var_line_list_part.append(allele_s5+\"|\"+str(1-int(allele_s5))+\":.:.:.:.:.:.:.:.:1\")\n",
    "        else:\n",
    "            var_line_list_part.append(\"./.:.:.:.:.:.:.:.:.:.\")\n",
    "            \n",
    "        var_lines_list_imp.append(var_line_list_part)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1|0',\n",
       " '855',\n",
       " '738',\n",
       " '24717',\n",
       " '117',\n",
       " '3956',\n",
       " '-245.824,0,-2114.31',\n",
       " '16147398',\n",
       " '.',\n",
       " '.']"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_line_list_imp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(var_line_list_imp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['0|1',\n",
       " '104',\n",
       " '62',\n",
       " '2231',\n",
       " '42',\n",
       " '1554',\n",
       " '-138.29,0,-199.21',\n",
       " '16050159',\n",
       " '.',\n",
       " '.']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_lines_list_imp[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GT:DP:RO:QR:AO:QA:GL:PS:PQ:PD'"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gt_flags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GT:DP:RO:QR:AO:QA:GL:PS'"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_line_list[8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1:50821579'"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matrix_row_list_prd[6][:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_out_address=vcf_file_address+\"_imp.vcf\"\n",
    "\n",
    "vcf_out_file=open(vcf_out_address,'w');\n",
    "\n",
    "for header_line in header_lines_list:\n",
    "    vcf_out_file.write(header_line)\n",
    "    \n",
    "for var_line_list in var_lines_list_imp:\n",
    "    \n",
    "    vcf_out_file.write(\"\\t\".join(var_line_list)+\"\\n\")\n",
    "\n",
    "vcf_out_file.close()\n",
    "   \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['22',\n",
       " '51243297',\n",
       " '.',\n",
       " 'A',\n",
       " 'T',\n",
       " '184.095',\n",
       " '.',\n",
       " 'DPB=10;EPPR=3.0103;GTI=0;MQMR=13.5;NS=1;NUMALT=1;ODDS=7.93999;PAIREDR=0.5;PQR=0;PRO=0;QR=26;RO=2;RPPR=7.35324;SRF=1;SRP=3.0103;SRR=1;DP=40;AB=0;ABP=0;AF=1;AO=8;CIGAR=1X;DPRA=0;EPP=12.7819;LEN=1;MEANALT=1;MQM=25.25;PAIRED=1;PAO=0;PQA=0;QA=275;RPL=7;RPP=12.7819;RPR=1;RUN=1;SAF=0;SAP=20.3821;SAR=8;TYPE=snp;technology.illumina=1;AN=8;AC=8',\n",
       " 'GT:DP:RO:QR:AO:QA:GL:PS']"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "var_line_list_part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'varpos_list_inmatrix' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-35-ef391acc53ab>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m#                                   str(allele_s_p)+\":\"+str(blockid_s_p)]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m     \u001b[0mvarpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvar_line_list_part\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 11\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mvarpos\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvarpos_list_inmatrix\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     12\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     13\u001b[0m         \u001b[0mgt_flags\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvar_line_list\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m8\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'varpos_list_inmatrix' is not defined"
     ]
    }
   ],
   "source": [
    "var_lines_list_hifi = []\n",
    "for var_line_list in var_lines_list: # \n",
    "    var_line_list_part=var_line_list[:9]\n",
    "    \n",
    "        \n",
    "#     matrix_row_list_prd = [str(varpos),str(allele_s1)+\":\"+str(blockid_s1),\n",
    "#                                   str(allele_s2)+\":\"+str(blockid_s2),str(allele_s3)+\":\"+str(blockid_s3),\n",
    "#                                   str(allele_s4)+\":\"+str(blockid_s4), str(allele_s5),\n",
    "#                                   str(allele_s_p)+\":\"+str(blockid_s_p)] \n",
    "    varpos = var_line_list_part[1]\n",
    "    if varpos in varpos_list_inmatrix:\n",
    "        \n",
    "        gt_flags = var_line_list[8]\n",
    "        idx_ps = gt_flags.split(\":\").index(\"PS\")\n",
    "        idx_gt = gt_flags.split(\":\").index(\"GT\")        \n",
    "        s1 = var_line_list[9] # the rest of var info is got from the s1\n",
    "        split_s1 = s1.split(\":\")\n",
    "\n",
    "        blockid_s1 = split_s1[idx_ps]\n",
    "        allele_s1 = split_s1[idx_gt]\n",
    "        \n",
    "        if allele_s1 != \"./.\":\n",
    "\n",
    "            var_idx = varpos_list_inmatrix.index(varpos)\n",
    "            matrix_row_list_prd = matrix_row_lists_prd[var_idx]   \n",
    "\n",
    "            var_line_list_imp = split_s1\n",
    "            var_line_list_part.append(\":\".join(var_line_list_imp))\n",
    "\n",
    "            allele_s5 = matrix_row_list_prd[5]\n",
    "            if allele_s5 != \".\":\n",
    "                var_line_list_part.append(allele_s5+\"|\"+str(1-int(allele_s5))+\":.:.:.:.:.:.:.:.:1\")\n",
    "            else:\n",
    "                var_line_list_part.append(\"./.:.:.:.:.:.:.:.:.:.\")\n",
    "\n",
    "            var_lines_list_hifi.append(var_line_list_part)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_out_address=vcf_file_address+\"_hifi.vcf\"\n",
    "\n",
    "vcf_out_file=open(vcf_out_address,'w');\n",
    "\n",
    "for header_line in header_lines_list:\n",
    "    vcf_out_file.write(header_line)\n",
    "    \n",
    "for var_line_list in var_lines_list_hifi:\n",
    "    \n",
    "    vcf_out_file.write(\"\\t\".join(var_line_list)+\"\\n\")\n",
    "\n",
    "vcf_out_file.close()\n",
    "   \n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "diff_list=[]\n",
    "\n",
    "for block_i in range(len(allele_blocks_s1)): # \n",
    "   \n",
    "    allele_block_s1=allele_blocks_s1[block_i]\n",
    "    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "\n",
    "    for var_i, varpos in enumerate(varpos_block_s1):\n",
    "        allele_s1= allele_block_s1[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "        \n",
    "            if allele_s1!=allele_s5:\n",
    "                \n",
    "                for block_i in range(number_phaseblocks_s2): # \n",
    "   \n",
    "                    allele_block_s2=allele_blocks_s2[block_i]\n",
    "                    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "                    if varpos in varpos_block_s2:\n",
    "                        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "                        diff_list.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                        #print([varpos,allele_s1,allele_s2,allele_s5],\"\\n\")\n",
    "                        break\n",
    "\n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "                \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-38-ba2a5007d8cd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m                     \u001b[0mvarpos_block_s1\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvarpos_blocks_s1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mblock_i\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m                     \u001b[0;32mif\u001b[0m \u001b[0mvarpos\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvarpos_block_s1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m                         \u001b[0mallele_s1\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mallele_block_s1\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvarpos_block_s2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvarpos\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m                         \u001b[0mdiff_list_s2\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvarpos\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mallele_s1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mallele_s2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mallele_s5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m                         \u001b[0;31m#print([varpos,allele_s1,allele_s2,allele_s5],\"\\n\")\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "\n",
    "diff_list_s2=[]\n",
    "\n",
    "for block_i in range(len(allele_blocks_s2)): # \n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "\n",
    "    for var_i, varpos in enumerate(varpos_block_s2):\n",
    "        allele_s2= allele_block_s2[var_i]\n",
    "        \n",
    "        if varpos in varposs_s5:\n",
    "            allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "        \n",
    "            if allele_s2!=allele_s5:\n",
    "                \n",
    "                for block_i in range(number_phaseblocks_s1): # \n",
    "   \n",
    "                    allele_block_s1=allele_blocks_s1[block_i]\n",
    "                    varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "                    if varpos in varpos_block_s1:\n",
    "                        allele_s1= allele_block_s1[varpos_block_s2.index(varpos)]\n",
    "                        diff_list_s2.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                        #print([varpos,allele_s1,allele_s2,allele_s5],\"\\n\")\n",
    "                        break\n",
    "\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hifiok_illno=[]\n",
    "hifino_illno=[]\n",
    "hifino_illok=[]\n",
    "\n",
    "for [varpos, hifi,illu,parnt] in diff_list_s2:\n",
    "    if illu!=parnt and hifi==parnt:\n",
    "        hifiok_illno.append(varpos)\n",
    "    if illu!=parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illno.append(varpos)       \n",
    "    if illu==parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illok.append(varpos) \n",
    "    \n",
    "print(len(diff_list_s2),len(hifiok_illno),len(hifino_illok),len(hifino_illno))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hifiok_illno=[]\n",
    "hifino_illno=[]\n",
    "hifino_illok=[]\n",
    "\n",
    "for [varpos, hifi,illu,parnt] in diff_list:\n",
    "    \n",
    "    if illu!=parnt and hifi==parnt:\n",
    "        hifiok_illno.append(varpos)\n",
    "    if illu!=parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illno.append(varpos)       \n",
    "    if illu==parnt and hifi!=parnt:\n",
    "        #print([varpos, hifi,illu,parnt] )\n",
    "        hifino_illok.append(varpos) \n",
    "    \n",
    "print(len(diff_list_s2),len(hifiok_illno),len(hifino_illok),len(hifino_illno))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_s5==allele_s2\n",
    "[allele_s1,allele_s5,allele_s2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31913 5561 2474 1302 1172\n",
      "15980 15349 631\n"
     ]
    }
   ],
   "source": [
    "similiar_count=0\n",
    "dissimiliar_count=0\n",
    "dissimiliar_count_s1ok=0\n",
    "dissimiliar_count_s2ok=0\n",
    "dissimiliar_count_exist5=0\n",
    "similiar_count_s1s2ok=0\n",
    "similiar_count_s1s2no=0\n",
    "similiar_count_exist5=0\n",
    "\n",
    "for block_i in range(number_phaseblocks_s2): # \n",
    "   \n",
    "    allele_block_s2=allele_blocks_s2[block_i]\n",
    "    varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "\n",
    "    intersected_idins1_s2=intersected_blocks_idins1_s2[block_i]\n",
    "    if intersected_idins1_s2 !=0:\n",
    "        block_indx_s1= blockid_blocks_s1.index(intersected_idins1_s2)\n",
    "    varpos_block_s1=varpos_blocks_s1[block_indx_s1]\n",
    "    allele_block_s1=allele_blocks_s1[block_indx_s1]\n",
    "    \n",
    "    intersect_varpos=set(varpos_block_s1) & set(varpos_block_s2)\n",
    "\n",
    "    for varpos in intersect_varpos:\n",
    "        allele_s1= allele_block_s1[varpos_block_s1.index(varpos)]\n",
    "        allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "        if allele_s1==allele_s2: \n",
    "            similiar_count+=1\n",
    "            if varpos in varposs_s5:\n",
    "                allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "                similiar_count_exist5+=1\n",
    "                if allele_s1==allele_s5 and allele_s2==allele_s5:\n",
    "                    similiar_count_s1s2ok+=1\n",
    "                if allele_s1!=allele_s5 and allele_s2!=allele_s5:\n",
    "                    similiar_count_s1s2no+=1\n",
    "        else:\n",
    "            #print(varpos)\n",
    "            dissimiliar_count+=1\n",
    "            if varpos in varposs_s5:\n",
    "                allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "                dissimiliar_count_exist5+=1\n",
    "                if allele_s1==allele_s5 and allele_s2!=allele_s5:\n",
    "                    dissimiliar_count_s1ok+=1\n",
    "                if allele_s1!=allele_s5 and allele_s2==allele_s5:\n",
    "                    dissimiliar_count_s2ok+=1\n",
    "                    \n",
    "\n",
    "            \n",
    "#     if similiar_count>1 and dissimiliar_count>1:\n",
    "#         print(similiar_count)\n",
    "\n",
    "print(similiar_count,dissimiliar_count,dissimiliar_count_exist5,dissimiliar_count_s1ok,dissimiliar_count_s2ok)\n",
    "print(similiar_count_exist5,similiar_count_s1s2ok,similiar_count_s1s2no)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(allele_block_s2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blockid  16099924\n",
      "Pos HiFi Illm parental \n",
      "\n",
      "16101684 1 1 0 \n",
      "\n",
      "16104385 1 1 0 \n",
      "\n",
      "16105717 1 1 0 \n",
      "\n",
      "16107307 1 1 0 \n",
      "\n",
      "16107662 1 1 0 \n",
      "\n",
      "16108260 1 1 0 \n",
      "\n",
      "16113547 0 1 1 \n",
      "\n",
      "16117006 0 1 1 \n",
      "\n",
      "16120074 1 1 0 \n",
      "\n",
      "16123812 1 1 0 \n",
      "\n",
      "16137152 1 1 0 \n",
      "\n",
      "16142002 1 1 0 \n",
      "\n",
      "16143370 1 1 0 \n",
      "\n",
      "16143995 1 1 0 \n",
      "\n",
      "16144768 1 1 0 \n",
      "\n",
      "16144952 1 1 0 \n",
      "\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "\n",
    "diff_list=[]\n",
    "count_simi=0\n",
    "block_i=1\n",
    "print(\"blockid \",blockid_blocks_s1[block_i])\n",
    "\n",
    "allele_block_s1=allele_blocks_s1[block_i]\n",
    "varpos_block_s1=varpos_blocks_s1[block_i]\n",
    "\n",
    "print(\"Pos\",\"HiFi\",\"Illm\",\"parental\",\"\\n\")\n",
    "for var_i, varpos in enumerate(varpos_block_s1):\n",
    "    allele_s1= allele_block_s1[var_i]\n",
    "\n",
    "    if varpos in varposs_s5:\n",
    "        allele_s5= alleles_s5[varposs_s5.index(varpos)]\n",
    "\n",
    "        if allele_s1!=allele_s5:\n",
    "\n",
    "            for block_i in range(number_phaseblocks_s2): # \n",
    "\n",
    "                allele_block_s2=allele_blocks_s2[block_i]\n",
    "                varpos_block_s2=varpos_blocks_s2[block_i]\n",
    "                if varpos in varpos_block_s2:\n",
    "                    allele_s2= allele_block_s2[varpos_block_s2.index(varpos)]\n",
    "                    diff_list.append([varpos,allele_s1,allele_s2,allele_s5])\n",
    "                    print(varpos,allele_s1,allele_s2,allele_s5,\"\\n\")\n",
    "                    break\n",
    "        if allele_s1==allele_s5:\n",
    "            count_simi+=1\n",
    "            \n",
    "\n",
    "\n",
    "                \n",
    "print(count_simi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum1=0\n",
    "for  a in varpos_blocks_s1:\n",
    "    sum1+=len(a)\n",
    "print(sum1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum2=0\n",
    "for  a in varpos_blocks_s2:\n",
    "    sum2+=len(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# coverage check\n",
    "\n",
    "bed_address=\"/Volumes/work/myjupyter/phaseyou/depth.regions.bed\"\n",
    "\n",
    "bed_file=open(bed_address,'r');\n",
    "\n",
    "selected_lines=[]\n",
    "start_pos_list=[]\n",
    "depth_list=[]\n",
    "\n",
    "for line in bed_file:\n",
    "    line_strip=line.strip()\n",
    "    line_split=line_strip.split(\"\\t\")\n",
    "    \n",
    "    \n",
    "    depth=float(line_split[3])\n",
    "    depth_list.append(depth)\n",
    "    \n",
    "    pos=float(line_split[1])\n",
    "    start_pos_list.append(pos)\n",
    "    \n",
    "    \n",
    "bed_file.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16.490618 29.0\n",
      "16.506873 22.0\n",
      "16.528995 34.0\n",
      "16.553221 24.0\n",
      "16.558766 26.0\n",
      "16.561186 34.0\n",
      "16.564036 34.0\n",
      "16.664149 20.0\n",
      "16.882859 94.0\n",
      "16.883072 92.0\n",
      "16.883221 92.0\n",
      "16.88483 88.0\n",
      "18.764533 70.0\n",
      "18.770724 75.0\n",
      "18.841654 55.0\n",
      "18.8419 55.0\n",
      "18.935759 42.0\n",
      "18.935842 42.0\n",
      "19.014037 28.0\n",
      "21.468861 27.0\n",
      "39.072681 27.0\n",
      "48.600367 30.0\n",
      "51.146139 26.0\n"
     ]
    }
   ],
   "source": [
    "list_pos=[16490618,16506873,16528995,16553221,16558766,16561186,16564036,16664149,16882859,16883072,16883221,16884830,18764533,18770724,18841654,\n",
    "       18841900,18935759,18935842,19014037,21468861,39072681,48600367,51146139]\n",
    "\n",
    "list1_depth=[]\n",
    "for  pos1 in list_pos:\n",
    "    depth_pos1=depth_list[int(pos1/500)]\n",
    "    list1_depth.append(depth_pos1)\n",
    "    print(pos1/1000000, depth_pos1)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "26.0"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "np.mean(depth_pos1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEGCAYAAABiq/5QAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAQYklEQVR4nO3df6zddX3H8ecLqhPqmEUupAKuShiOEBVtFMfinBDj1AhzwjTDFMdkS2SCczo0SwoxywhR45JNkoqyqkxTAQNjibOrusVtQVpAoVQHkR8Da3vROcUwFXnvj/OtXG7vbU9rv/d8D5/nI7k55/s959z7yift63zP53x/pKqQJLXjoEkHkCQtLYtfkhpj8UtSYyx+SWqMxS9JjVk26QDjOOKII2rVqlWTjiFJU2XLli0PVdXM/PVTUfyrVq1i8+bNk44hSVMlyX0LrXeqR5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGjMVR+4eSJcmP7+/1ovQSGqQW/yS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5Ia02vxJ3lnkq1J7kjy6SRPS3J4ko1J7upuV/SZQZL0RL0Vf5KjgXcAq6vqJOBg4E3AxcCmqjoe2NQtS5KWSN9TPcuAQ5IsAw4Fvg2cAazvHl8PnNlzBknSHL0Vf1U9CHwAuB/YDvxvVX0BOKqqtnfP2Q4c2VcGSdLu+pzqWcFo6/45wLOA5UnO2YfXn59kc5LNs7OzfcWUpOb0OdVzOnBPVc1W1U+B64DfAHYkWQnQ3e5c6MVVta6qVlfV6pmZmR5jSlJb+iz++4FTkhyaJMBpwDbgBmBN95w1wPU9ZpAkzbOsr19cVTcluQa4BXgUuBVYBzwd2JDkPEZvDmf1lUGStLveih+gqtYCa+et/jGjrX9J0gR45K4kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqTK+7cz4ZXZr8/P7aqgkmkaT94xa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGtNr8Sd5RpJrknwjybYkL0tyeJKNSe7qblf0mUGS9ER9b/H/DfD5qnoe8AJgG3AxsKmqjgc2dcuSpCXSW/EnOQx4OfAxgKr6SVV9HzgDWN89bT1wZl8ZJEm763OL/7nALHBVkluTXJlkOXBUVW0H6G6P7DGDJGmePot/GfAi4IqqOhn4EfswrZPk/CSbk2yenZ3tK6MkNafP4n8AeKCqbuqWr2H0RrAjyUqA7nbnQi+uqnVVtbqqVs/MzPQYU5La0lvxV9V3gP9OckK36jTgTuAGYE23bg1wfV8ZJEm7WzbOk5IcCrwLeHZVvS3J8cAJVXXjXl76p8DVSZ4KfAt4K6M3mw1JzgPuB87a7/SSpH02VvEDVwFbgJd1yw8AnwX2WPxVdRuweoGHThvz70qSDrBxp3qOq6rLgZ8CVNUjQHpLJUnqzbjF/5MkhwAFkOQ44Me9pZIk9WbcqZ61wOeBY5NcDZwKnNtXKElSf8Yq/qramOQW4BRGUzwXVtVDvSaTJPVi3L16XtTd3d7dPjvJrwD3VdWjvSSTJPVi3KmejzA6+OrrjLb4T+ruPzPJn1TVF3rKJ0k6wMb9cvde4OTuSNoXAycDdwCnA5f3lG3wLo07NkmaPuMW//Oqauuuhaq6k9Ebwbf6iSVJ6su4Uz3fTHIF8Jlu+feB/0ryS3T79kuSpsO4W/znAncDFwHvZHT6hXMZlf5v95BLktSTcXfnfAT4YPcz38MHNJEkqVfj7s55PPDXwInA03atr6rn9pRLktSTcad6rgKuAB5lNLXzCeCTfYWSJPVn3OI/pKo2Aamq+6rqEuCV/cWSJPVl3L16/i/JQcBdSS4AHsRr5UrSVBp3i/8i4FDgHcCLgXN4/CpakqQpstct/iQHA2dX1bsZ7cHz1t5TSZJ6s9ct/qr6GfDixPMTSNKTwbhz/LcC1yf5LPCjXSur6rpeUkmSejNu8R8OfJcn7slTgMUvSVNm3CN3ndeXpCeJsfbqSfJrSTYluaNbfn6Sv+w3miSpD+PuzvlR4L10Z+Ksqq8Db+orlCSpP+MW/6FV9dV567zkoiRNoXGL/6EkxzH6Qpckb+Tx6+9KkqbIuHv1vB1YBzwvyYPAPcAf9JZKktSbcYv/vqo6Pcly4KCq+mGfoSRJ/Rl3queeJOuAU/DCK5I01cYt/hOAf2E05XNPkr9N8pv9xZIk9WWs4q+qR6pqQ1W9ATgZOAz4116TSZJ6Me4WP0l+K8lHgFsYXX7x7N5SSZJ6M+41d+8BbgM2AO+uqh/t+RWSpKEad6+eF1TVD3pNIklaEuNO9RyW5HNJdibZkeTaJMf0mkyS1Itxi/8q4AbgWcDRwD9265p3qdenkTRlxi3+maq6qqoe7X7+HpjpMZckqSf7cq6ec5Ic3P2cw+jCLHvVPf/WJDd2y4cn2Zjkru52xf6GlyTtu3GL/w8Z7b75HUYnZ3sj4190/UJg25zli4FNVXU8sKlbliQtkXGL//3AmqqaqaojGb0RXLK3F3VfAL8WuHLO6jOA9d399cCZ44aVJP3ixi3+51fV/+xaqKrvMTqCd28+DLwHeGzOuqOqanv3e7YDRy70wiTnJ9mcZPPs7OyYMSVJezNu8R80dy4+yeHs5RiAJK8DdlbVlv0JVlXrqmp1Va2emfF7ZEk6UMY9gOuDwH8kuYbRxVjOBv5qL685FXh9ktcwOsXDYUk+BexIsrKqtidZCezcz+ySpP0w7knaPgH8HrADmAXeUFWf3Mtr3ltVx1TVKkbX5/1iVZ3D6HiANd3T1gDX72d2SdJ+GHeLn6q6E7jzAPzNy4ANSc4D7gfOOgC/U5I0prGL/xdRVV8Gvtzd/y5w2lL8XUnS7sY+LbMk6cnB4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNWZL9+CfNq2RJ0uPc4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1pokDuCTpQJt7YOjaqgkm2Xdu8UtSY570xe/pGiTpiZ70xS9JeiKL/wDwU4WkaWLxS1Jj3KtnQOZ/cpi2PQU0fP4bE7jFL0nNsfglqTFO9UjSgCzFdJxb/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1Jjeij/JsUm+lGRbkq1JLuzWH55kY5K7utsVfWWQJO2uzy3+R4F3VdWvA6cAb09yInAxsKmqjgc2dcuSpCXSW/FX1faquqW7/0NgG3A0cAawvnvaeuDMvjJIkna3JHP8SVYBJwM3AUdV1XYYvTkARy7ymvOTbE6yeXZ2diliSlITei/+JE8HrgUuqqofjPu6qlpXVauravXMzEx/ASWpMb0Wf5KnMCr9q6vqum71jiQru8dXAjv7zCBJeqI+9+oJ8DFgW1V9aM5DNwBruvtrgOv7yiBJ2l2fF2I5FXgLcHuS27p17wMuAzYkOQ+4HzirxwySpHl6K/6q+gqQRR4+ra+/K0naM4/claTGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGmPxS1JjLH5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhpj8UtSYyx+SWqMxS9JjbH4JakxFr8kNcbil6TGWPyS1BiLX5IaY/FLUmMsfklqjMUvSY2x+CWpMRa/JDXG4pekxlj8ktQYi1+SGjOR4k/y6iTfTHJ3kosnkUGSWrXkxZ/kYODvgN8BTgTenOTEpc4hSa2axBb/S4C7q+pbVfUT4DPAGRPIIUlNSlUt7R9M3gi8uqr+qFt+C/DSqrpg3vPOB87vFk8Avjnn4SOAh5Yg7oEwTVlhuvJOU1aYrrxm7c9S5v3VqpqZv3LZEv3xubLAut3efapqHbBuwV+QbK6q1Qc6WB+mKStMV95pygrTldes/RlC3klM9TwAHDtn+Rjg2xPIIUlNmkTx3wwcn+Q5SZ4KvAm4YQI5JKlJSz7VU1WPJrkA+GfgYODjVbV1H3/NglNAAzVNWWG68k5TVpiuvGbtz8TzLvmXu5KkyfLIXUlqjMUvSY0ZfPEn+XiSnUnumLPukiQPJrmt+3nNJDPukuTYJF9Ksi3J1iQXdusPT7IxyV3d7YoBZx3q2D4tyVeTfK3Le2m3fohju1jWQY4tjI6oT3Jrkhu75cGN6y4LZB3yuN6b5PYu1+Zu3cTHdvBz/EleDjwMfKKqTurWXQI8XFUfmGS2+ZKsBFZW1S1JfhnYApwJnAt8r6ou685NtKKq/mJySfeY9WyGObYBllfVw0meAnwFuBB4A8Mb28WyvpoBji1Akj8DVgOHVdXrklzOwMZ1lwWyXsJwx/VeYHVVPTRn3cTHdvBb/FX1b8D3Jp1jHFW1vapu6e7/ENgGHM3olBTru6etZ1SwE7WHrINUIw93i0/pfophju1iWQcpyTHAa4Er56we3LjColmnzcTHdvDFvwcXJPl6NxU0mI+huyRZBZwM3AQcVVXbYVS4wJETjLabeVlhoGPbfcS/DdgJbKyqwY7tIllhmGP7YeA9wGNz1g1yXFk4KwxzXGH0hv+FJFsyOg0NDGBsp7X4rwCOA14IbAc+ONE08yR5OnAtcFFV/WDSefZkgayDHduq+llVvZDR0d4vSXLShCMtapGsgxvbJK8DdlbVlkln2Zs9ZB3cuM5xalW9iNHZiN/eTV1P3FQWf1Xt6P5jPQZ8lNEZPwehm9O9Fri6qq7rVu/o5tR3za3vnFS+uRbKOuSx3aWqvg98mdGc+SDHdpe5WQc6tqcCr+/moj8DvDLJpxjmuC6YdaDjCkBVfbu73Ql8jlG2iY/tVBb/rkHr/C5wx2LPXUrdl3ofA7ZV1YfmPHQDsKa7vwa4fqmzzbdY1gGP7UySZ3T3DwFOB77BMMd2waxDHNuqem9VHVNVqxidPuWLVXUOAxzXxbIOcVwBkizvdpwgyXLgVYyyTXxsJ3F2zn2S5NPAK4AjkjwArAVekeSFjObP7gX+eFL55jkVeAtweze/C/A+4DJgQ5LzgPuBsyYT7wkWy/rmgY7tSmB9RhfyOQjYUFU3JvlPhje2i2X95EDHdiFD/De7mMsHOq5HAZ8bbWOxDPiHqvp8kpuZ8NgOfndOSdKBNZVTPZKk/WfxS1JjLH5JaozFL0mNsfglqTEWv3QAdGeI/PNJ55DGYfFLUmMsfmkvkqxK8o0kVya5I8nVSU5P8u/dOdV3nSLgBUm+2K1720RDS3vgAVzSXnRnL72b0RlMtwI3A18DzgNeD7wVuI3R6QJOAZYDtwIv3XWuFmlI3OKXxnNPVd3enQhsK7CpRltNtwOruudcX1WPdBfd+BIDOlmYNJfFL43nx3PuPzZn+TEeP+fV/I/PfpzWIFn80oFzRkbX230moxML3jzhPNKCBn92TmmKfBX4J+DZwPud39dQ+eWuJDXGqR5JaozFL0mNsfglqTEWvyQ1xuKXpMZY/JLUGItfkhrz/+l2CUm/OB7PAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt  \n",
    "\n",
    "\n",
    "list_pos1=[i/1e6 for i in list_pos]\n",
    "plt.bar(list_pos1,list1_depth, color ='maroon', width = .4)\n",
    "#plt.show()\n",
    "\n",
    "plt.xlabel(\"mb\")\n",
    "\n",
    "plt.ylabel(\"coverage\")\n",
    "\n",
    "plt.savefig(\"Coverage_.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAl4AAAFNCAYAAADRi2EuAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAAi+ElEQVR4nO3deZglZX328e/tAIqAKEyDbCOIoMGFUQdcMFFRDBgVQVERfUGII29QUDFGja+ABuMGxCXRQNg0CIKK4i5BENxYhQEExLAo+yLIogIz/N4/qloPYy9neuZU95z5fq6rr3Pqqe13ThfTN089VZWqQpIkSYP3sOkuQJIkaUVh8JIkSeqIwUuSJKkjBi9JkqSOGLwkSZI6YvCSJEnqiMFLmsGSnJHk76e7jq4k2SPJj3qmK8kTpritcddN8p0ku0+1zsW2dUySf2nf/3WSK3rmPTHJz5PcnWTfJKsm+UaS3yU5aVnsX9LyxeAlLYUkz0vyk/YP6W+T/DjJVu28h4SI6TSTapkJqmqHqjp2ANs9q6qe2NP0buCMqlqjqj4FvBpYF1i7qnZZ1vufSJIXJLmuy31K+ksGL2mKkjwK+CbwaWAtYAPgIOC+6axrJkuy0nTX0LHHAZcuNv3Lqlq4pBsalu9uWD6HNFUGL2nqNgeoquOralFV/aGqvl9VC5L8FfA54DlJ7klyJ/zlqcMxTq1tl+TytgftM0B6d5hkzySXJbkjyfeSPK5nXiXZO8mV7fx/T2O8Wl6a5BftabDrk7xrvA86lf32fL4fJzksyW+BA5OsmeTzSW5Ncm2S9yeZ9N+iJA9P8okkv05yc5LPJVm1Z/4/JrkxyQ1J9pxkW3/6PYz+Dtpt35Hk6iQ7TLDu05Nc0H5vXwIe0TPvT71KSX4AvBD4TPu9Hw98AHhtO71Xn9/tPkmuBK5s216W5MIkd7a9rU/rWf6aJO9KsqA9hr6U5BFJVgO+A6zf7vueJOuP8dlWTXJI+3v5Xfu9rNrOe0WSS9v9ntEeV711PqFnuvf06wuSXJfkn5LcBBydZHaSb7bb+m2Ss0aPgSTrJ/lKe3xcnWTfnu1uneS8JHe1x8ChE/2epZnI4CVN3S+BRUmOTbJDkseMzqiqy4C9gZ9W1epV9ejJNpZkNvAV4P3AbOB/gW165r8SeB+wMzACnAUcv9hmXgZsBWwJvAb42wlqORJ4S1WtATwF+ME4dU1pvz3zngVcBawDHEzTQ7gm8Hjg+cD/Ad40/jfzJx+lCbtzgSfQ9DB+oK1xe+BdwHbAZsCL+9her2cBV9B87x8DjhwNj72SrAJ8DfgCTS/nScCrxtpgVW1L8129tf3edwU+DHypnT6yz+/2lW19WyR5BnAU8BZgbeA/gVOSPLxn+dcA2wObAE8D9qiqe4EdgBvafa9eVTeMUfYngGcCz20/37uBB5Ns3tb19rbObwPfaL+Pfjy23d7jgPnA/sB17bbWbb+DasPXN4CLaH6/LwLenmT0ePok8MmqehSwKXBin/uXZgyDlzRFVXUX8DyggCOAW5OckmTdKW7ypcAvqurLVfUA8G/ATT3z3wL8a1Vd1p6q+jAwt7eHBPhIVd1ZVb8GTqcJKeN5gOaP+aOq6o6qumCc5ZZ2vzdU1afbde8HXgu8t6rurqprgEOAN05QJ20IejPwjqr6bVXd3dbxunaR1wBHV9Ulbcg4cKLtjeHaqjqiqhYBxwLr0QSCxT0bWBn4t6p6oKq+DJy7hPvq1c93+6/tZ/4DzXfwn1V1dtvLeizNqe1n9yz/qaq6oap+SxNi5vZTSBt69gT2q6rr2+3/pKruo/mdfauqTm2PzU8Aq9IEtH48CBxQVfe1n+MBmu/4ce33eFY1Dw7eChipqg9W1f1VdRXNf1ujv+cHgCckmV1V91TVz/rcvzRjGLykpdD+wdyjqjak6TVanyYwTcX6wG96tl290zS9BZ9sT8/cCfyW5lTkBj3L9Aa13wOrT7C/V9GEvWuT/DDJc8ZZbmn32/sZZgOrANf2tF272LbGMgI8Eji/p47vtu2w2He32Pb78af6q+r37duxvrv1gevb381U99Wrn+928WNg/9Hl23U2ausatSTHQK/ZNKdN/3eMeevT8zmr6sG2rsl+b6Nurao/9kx/HPgV8P0kVyV5T9v+OJrTob2f7338OQTvRdPreXmSc5O8rM/9SzOGgxylZaSqLk9yDE0vBjQ9YYu7lyZAjHpsz/sbaf6IAn/q5dmoZ/5vgIOr6riplDdGvecCOyZZGXgrzWmbjRZfbin3u/i+b6PptXgc8Iu2bQ5w/STbuA34A/Dkqhpr2Yd8d+02B+FGYIMk6Qlfcxg7rPSjn++29/sbXf7gKexrrOOx123AH2lO4V202LwbgKeOTvQcm6O/i9/zl8d17xWUD9l322O5P02IfDJwepJzaT7f1VW12ZgfoOpKYNe2d25n4MtJ1m57OaXlgj1e0hQleVKS/ZNs2E5vBOwKjJ7+uBnYcLFxMBcCOyd5ZDsYea+eed8Cnpxk5zRXfu3LQ4PZ54D3tn+oSDNIvd9bEjykliSrJNktyZrtqaO7gEXjrLs0+32I9lTeicDBSdZoT6m9E/jvSdZ7kOaU02FJ1mnr2KBn7M+JwB5JtkjySOCAqdTXh58CC4F9k6yUZGdg66XY3pJ+t0cAeyd5VhqrJfm7JGv0sa+bgbWTrDnWzPY7Pgo4tB3gPivJc9rxYycCf5fkRW1Q35/mFOdP2tUvBF7frrM9zdi9caW5QOAJbYAbPfYWAecAd6UZiL9qu72n5M+3aHlDkpG21jvbzY133EozksFLmrq7aQY9n53kXprAdQnNHyVoBqtfCtyU5La27TCacU4304wl+lNPR1XdBuwCfAS4nWaQ+I975p9MM8D8hCR3tfsa9+q7xYxVyxuBa9pt7Q28YawVl3K/Y3kbTc/fVcCPgC/S/MGfzD/RnJ76WVvH/wBPbGv8Ds0p3h+0y4x5ocDSqqr7aXpa9gDuoBn79NWl2N4SfbdVdR7NOK/PtPv/VVtLP/u6nGaA/FXtaby/uKqR5gKFi2nGrf22re1hVXUFzfHxaZqesZcDL2+/D4D92rY7gd1oLkCYyGY0v797aMLsf1TVGW0wfznNuLSr2339F83FGNBcNHBpkntoBtq/brFTmNKMl4cOVZAkSdKg2OMlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1JHl4gaqs2fPro033ni6y5AkSZrU+eeff1tVjYw1b7kIXhtvvDHnnXfedJchSZI0qSTjPkrMU42SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1JGBBa8kj0hyTpKLklya5KC2/cAk1ye5sP156aBqkCRJmkkGeQPV+4Btq+qeJCsDP0rynXbeYVX1iQHuW5IkacYZWPCqqgLuaSdXbn9qUPuTJEma6QY6xivJrCQXArcAp1bV2e2styZZkOSoJI8ZZA2SJEkzRZqOqQHvJHk0cDLwNuBW4Daa3q8PAetV1Z5jrDMfmA8wZ86cZ1577biPPZIkSVN0UDLdJXTqgG5yz/lVNW+seZ1c1VhVdwJnANtX1c1VtaiqHgSOALYeZ53Dq2peVc0bGRnzAd+SJEnLlUFe1TjS9nSRZFXgxcDlSdbrWWwn4JJB1SBJkjSTDPKqxvWAY5PMogl4J1bVN5N8IclcmlON1wBvGWANkiRJM8Ygr2pcADx9jPY3DmqfkiRJM5l3rpckSeqIwUuSJKkjBi9JkqSOGLwkSZI6YvCSJEnqiMFLkiSpIwYvSZKkjhi8JEmSOmLwkiRJ6ojBS5IkqSMGL0mSpI4YvCRJkjpi8JIkSeqIwUuSJKkjBi9JkqSOGLwkSZI6YvCSJEnqiMFLkiSpIwYvSZKkjhi8JEmSOmLwkiRJ6ojBS5IkqSMGL0mSpI4YvCRJkjpi8JIkSeqIwUuSJKkjBi9JkqSOGLwkSZI6MrDgleQRSc5JclGSS5Mc1LavleTUJFe2r48ZVA2SJEkzySB7vO4Dtq2qLYG5wPZJng28BzitqjYDTmunJUmSht7Aglc17mknV25/CtgROLZtPxZ45aBqkCRJmkkGOsYryawkFwK3AKdW1dnAulV1I0D7us4ga5AkSZopVhrkxqtqETA3yaOBk5M8pd91k8wH5gPMmTNnMAVK0nLmoGS6S+jMAVXTXYK0zHVyVWNV3QmcAWwP3JxkPYD29ZZx1jm8quZV1byRkZEuypQkSRqoQV7VONL2dJFkVeDFwOXAKcDu7WK7A18fVA2SJEkzySBPNa4HHJtkFk3AO7Gqvpnkp8CJSfYCfg3sMsAaJEmSZoyBBa+qWgA8fYz224EXDWq/kiRJM5V3rpckSeqIwUuSJKkjBi9JkqSOGLwkSZI6YvCSJEnqiMFLkiSpIwYvSZKkjhi8JEmSOmLwkiRJ6ojBS5IkqSMGL0mSpI4YvCRJkjpi8JIkSeqIwUuSJKkjBi9JkqSOGLwkSZI6YvCSJEnqiMFLkiSpIwYvSZKkjhi8JEmSOmLwkiRJ6ojBS5IkqSMGL0mSpI6sNN6MJBcDNd78qnraQCqSJEkaUuMGL+Bl7es+7esX2tfdgN8PrCJJkqQhNW7wqqprAZJsU1Xb9Mx6T5IfAx8cdHGSJEnDpJ8xXqsled7oRJLnAqsNriRJkqThNNGpxlF7AUclWZNmzNfvgD0HWpUkSdIQmrTHq6rOr6otgacBc6tqblVdMNl6STZKcnqSy5JcmmS/tv3AJNcnubD9eenSfwxJkqSZb9IeryTrAh8G1q+qHZJsATynqo6cZNWFwP5VdUGSNYDzk5zazjusqj6xVJVLkiQtZ/oZ43UM8D1g/Xb6l8DbJ1upqm4c7RmrqruBy4ANplSlJEnSEOgneM2uqhOBBwGqaiGwaEl2kmRj4OnA2W3TW5MsSHJUkscsybYkSZKWV/0Er3uTrE17M9Ukz6YZYN+XJKsDXwHeXlV3AZ8FNgXmAjcCh4yz3vwk5yU579Zbb+13d5IkSTNWP8HrncApwKbt/bs+D7ytn40nWZkmdB1XVV8FqKqbq2pRVT0IHAFsPda6VXV4Vc2rqnkjIyP97E6SJGlGm3RwfTs4/vnAE4EAV1TVA5OtlyTAkcBlVXVoT/t6VXVjO7kTcMmUKpckSVrOTPSsxp3HmbV5EkZ7sCawDfBG4OIkF7Zt7wN2TTKX5tTlNcBblqRgSZKk5dVEPV4vn2BeARMGr6r6EU0P2eK+3UddkiRJQ2eiZzW+qctCJEmSht1EpxrfOdGKveO2JEmSNLmJTjWu0VkVkiRJK4CJTjUe1GUhkiRJw27S+3gl2TDJyUluSXJzkq8k2bCL4iRJkoZJPzdQPZrmBqrr0zxr8RttmyRJkpZAP8FrpKqOrqqF7c8xgLeSlyRJWkL9BK/bkrwhyaz25w3A7YMuTJIkadj0E7z2BF4D3ETzUOtXt22SJElaAhM+qzHJLODDVfWKjuqRJEkaWhP2eFXVImAkySod1SNJkjS0Juzxal0D/DjJKcC9o43euV6SJGnJ9BO8bmh/HoZ3s5ckSZqySYOXd7CXJElaNiYNXkk2B94FbNy7fFVtO7iyJEmShk8/pxpPAj4H/BewaLDlSJIkDa9+gtfCqvrswCuRJEkacv3cQPUbSf4hyXpJ1hr9GXhlkiRJQ6afHq/d29d/7Gkr4PHLvhxJkqTh1c9VjZt0UYgkSdKwm/RUY5JHJnl/ksPb6c2SvGzwpUmSJA2XfsZ4HQ3cDzy3nb4O+JeBVSRJkjSk+glem1bVx4AHAKrqD0AGWpUkSdIQ6id43Z9kVZoB9STZFLhvoFVJkiQNoX6uajwA+C6wUZLjgG2APQZZlCRJ0jDq56rGU5NcADyb5hTjflV128ArkyRJGjL99HhRVbcD3xpwLZIkSUOtnzFekiRJWgYGFrySbJTk9CSXJbk0yX5t+1pJTk1yZfv6mEHVIEmSNJP0cwPVTZM8vH3/giT7Jnl0H9teCOxfVX9FMz5snyRbAO8BTquqzYDT2mlJkqSh10+P11eARUmeABwJbAJ8cbKVqurGqrqgfX83cBmwAbAjcGy72LHAK5e8bEmSpOVPP8HrwapaCOwE/FtVvQNYb0l2kmRj4OnA2cC6VXUjNOEMWGeJKpYkSVpO9XNV4wNJdgV2B17etq3c7w6SrE7Ta/b2qror6e+m90nmA/MB5syZ0+/upuygPusaFgdUTXcJkiStcPrp8XoT8Bzg4Kq6OskmwH/3s/EkK9OEruOq6qtt881J1mvnrwfcMta6VXV4Vc2rqnkjIyP97E6SJGlG6yd4bVdV+1bV8QBVdTXwh8lWStO1dSRwWVUd2jPrFJreM9rXry9ZyZIkScunfoLX7mO07dHHetsAbwS2TXJh+/NS4CPAdkmuBLZrpyVJkobeuGO82nFdrwc2SXJKz6w1gNsn23BV/YjmEUNjedGSFClJkjQMJhpc/xPgRmA2cEhP+93AgkEWJUmSNIzGDV5VdS1wLc3AekmSJC2lfu5cv3P7eJ/fJbkryd1J7uqiOEmSpGHSz328Pga8vKouG3QxkiRJw6yfqxpvNnRJkiQtvX56vM5L8iXga8B9o409N0SVJElSH/oJXo8Cfg+8pKetAIOXJEnSEpg0eFXVm7ooRJIkadj1c1Xj5klOS3JJO/20JO8ffGmSJEnDpZ/B9UcA7wUeAKiqBcDrBlmUJEnSMOoneD2yqs5ZrG3hIIqRJEkaZv0Er9uSbEozoJ4kr6Z5lJAkSZKWQD9XNe4DHA48Kcn1wNXAGwZalSRJ0hDq56rGq4AXJ1kNeFhV3T34siRJkobPuMEryTvHaQegqg4dUE2SJElDaaIerzXa1ycCWwGntNMvB84cZFGSJEnDaNzgVVUHAST5PvCM0VOMSQ4ETuqkOkmSpCHSz1WNc4D7e6bvBzYeSDWSJElDrJ+rGr8AnJPkZJpbSuwEfH6gVUmSJA2hfq5qPDjJd4C/bpveVFU/H2xZkiRJw2fS4JVkDnAbcHJvW1X9epCFSZIkDZt+TjV+i/au9cCqwCbAFcCTB1WUJEnSMOrnVONTe6eTPAN4y8AqkiRJGlL9XNX4EFV1Ac19vSRJkrQE+hnj1XsH+4cBzwBuHVhFkiRJQ6qfMV5r9LxfSDPm6yuDKUeSJGl49RO8flFVD7lTfZJd8O71kiRJS6SfMV7v7bNNkiRJExi3xyvJDsBLgQ2SfKpn1qNoTjlOKMlRwMuAW6rqKW3bgcCb+fMYsfdV1benVrokSdLyZaIerxuA84A/Auf3/JwC/G0f2z4G2H6M9sOqam77Y+iSJEkrjHF7vKrqIuCiJF+sqgcAkjwG2Kiq7phsw1V1ZpKNl1mlkiRJy7l+xnidmuRRSdYCLgKOTnLoUuzzrUkWJDmqDXKSJEkrhH6ualyzqu5K8vfA0VV1QJIFU9zfZ4EP0TyC6EPAIcCeYy2YZD4wH2DOnDlT3J00vQ5KpruEzhxQNflCkrSC66fHa6Uk6wGvAb65NDurqpuralFVPQgcAWw9wbKHV9W8qpo3MjKyNLuVJEmaEfoJXh8Evgf8qqrOTfJ44Mqp7KwNcKN2Ai6ZynYkSZKWR/08JPskem6WWlVXAa+abL0kxwMvAGYnuQ44AHhBkrk0pxqvwYdtS5KkFUg/Y7ympKp2HaP5yEHtT5Ikaabr51SjJEmSloFxg1eS/drXbborR5IkaXhN1OP1pvb1010UIkmSNOwmGuN1WZJrgJHF7tsVoKrqaQOtTJIkachM9MigXZM8luZWEq/oriRJkqThNOFVjVV1E7BlklWAzdvmK0af3ShJkqT+TXo7iSTPBz5Pc9+tABsl2b2qzhxwbZIkSUOln/t4HQq8pKquAEiyOXA88MxBFiZJkjRs+rmP18qjoQugqn4JrDy4kiRJkoZTPz1e5yU5EvhCO70bcP7gSpIkSRpO/QSv/wvsA+xLM8brTOA/BlmUJEnSMOrnIdn30YzzOnTw5UiSJA0vn9UoSZLUEYOXJElSRwxekiRJHZlS8Eoyf1kXIkmSNOym2uOVZVqFJEnSCmBKwauq/nNZFyJJkjTsJg1eSTZMcnKSW5PcnOQrSTbsojhJkqRh0k+P19HAKcB6wAbAN9o2SZIkLYF+gtdIVR1dVQvbn2OAkQHXJUmSNHT6CV63JXlDklntzxuA2wddmCRJ0rDpJ3jtCbwGuAm4EXh12yZJkqQl0M+zGn8NvKKDWiRJkobauMEryQcmWK+q6kMDqEeSJGloTdTjde8YbasBewFrAwYvSZKkJTBu8KqqQ0bfJ1kD2A94E3ACcMh460mSJGlsE47xSrIW8E5gN+BY4BlVdUcXhUmSJA2bca9qTPJx4FzgbuCpVXXgkoSuJEcluSXJJT1tayU5NcmV7etjlqp6SZKk5chEt5PYH1gfeD9wQ5K72p+7k9zVx7aPAbZfrO09wGlVtRlwWjstSZK0QphojNeUHqDds/6ZSTZerHlH4AXt+2OBM4B/Wpr9SJIkLS+WKlxNwbpVdSNA+7pOx/uXJEmaNl0Hr74lmZ/kvCTn3XrrrdNdjiRJ0lLrOnjdnGQ9gPb1lvEWrKrDq2peVc0bGfGZ3JIkafnXdfA6Bdi9fb878PWO9y9JkjRtBha8khwP/BR4YpLrkuwFfATYLsmVwHbttCRJ0gph0odkT1VV7TrOrBcNap+SJEkz2YwdXC9JkjRsDF6SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1BGDlyRJUkcMXpIkSR0xeEmSJHXE4CVJktQRg5ckSVJHDF6SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1BGDlyRJUkcMXpIkSR0xeEmSJHXE4CVJktQRg5ckSVJHDF6SJEkdMXhJkiR1xOAlSZLUkZWmY6dJrgHuBhYBC6tq3nTUIUmS1KVpCV6tF1bVbdO4f0mSpE55qlGSJKkj0xW8Cvh+kvOTzJ+mGiRJkjo1Xacat6mqG5KsA5ya5PKqOrN3gTaQzQeYM2fOdNQoSZK0TE1Lj1dV3dC+3gKcDGw9xjKHV9W8qpo3MjLSdYmSJEnLXOfBK8lqSdYYfQ+8BLik6zokSZK6Nh2nGtcFTk4yuv8vVtV3p6EOSZKkTnUevKrqKmDLrvcrSZI03bydhCRJUkcMXpIkSR0xeEmSJHXE4CVJktQRg5ckSVJHDF6SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1BGDlyRJUkcMXpIkSR0xeEmSJHXE4CVJktQRg5ckSVJHDF6SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1BGDlyRJUkcMXpIkSR0xeEmSJHXE4CVJktSRaQleSbZPckWSXyV5z3TUIEmS1LXOg1eSWcC/AzsAWwC7Jtmi6zokSZK6Nh09XlsDv6qqq6rqfuAEYMdpqEOSJKlT0xG8NgB+0zN9XdsmSZI01Faahn1mjLb6i4WS+cD8dvKeJFcMtKrpMxu4reudHpixfg0aEh5TGoTOjyuPqaE3zP9WPW68GdMRvK4DNuqZ3hC4YfGFqupw4PCuipouSc6rqnnTXYeGh8eUBsHjSsvainpMTcepxnOBzZJskmQV4HXAKdNQhyRJUqc67/GqqoVJ3gp8D5gFHFVVl3ZdhyRJUtem41QjVfVt4NvTse8ZaOhPp6pzHlMaBI8rLWsr5DGVqr8Y1y5JkqQB8JFBkiRJHTF4TZMkj01yQpL/TfKLJN9Osvl016WZy2NGg5LknumuQcuXJIuSXJjkkiQnJXnkBMtunOT1PdN7JPlMN5XOPAavaZAkwMnAGVW1aVVtAbwPWHd6K9NM1c8x0/5jduAk27lmkHVKWmH8oarmVtVTgPuBvSdYdmPg9RPMX6EYvKbHC4EHqupzow1VdWFVnTWNNWlm85jRQCVZPclpSS5IcnGSHdv2jyb5h57lDkyy/3jLa4V0FvCEJB9Kst9oY5KDk+wLfAT467aH7B3t7PWTfDfJlUk+1rPOru3xdEmSj/a039Nu76IkP0uy3HZUGLymx1OA86e7CC1XPGY0aH8EdqqqZ9AE/UPantYTgNf2LPca4KQJltcKJMlKwA7AxcCRwO5t+8No7tN5HPAe4Ky2h+ywdtW5NMfVU4HXJtkoyfrAR4Ft2/lbJXllu/xqwM+qakvgTODNA/9wAzItt5OQtGwkWRs4rZ1cC1il5x+qN1bVxUn+HdimbVs/yYXt+5Oq6uDOitVMF+DDSf4GeJDmGbrrVtXPk6zT/lEcAe6oql8nWXms5YGbpql+dWvVnn9LzgKOrKr7k9ye5Ok0x8LPq+r2cfL4aVX1O4Akv6B5xM7aNMMpbm3bjwP+BvgazenMb7brng9sN5BP1QGD1/S4FHj1dBeh5cqYx0xV3U7zf4Yk2QPYuKoOXGyZfUbfJ7mmquYOsE4tv3ajCVbPrKoH2vGAj2jnfZnm+HssTQ/YZMtr+P1hnH9L/gvYg+ZYOWqC9e/reb+IJo9M1GP6QP35/lejyy+XPNU4PX4APDzJn7pKk2yV5PnTWJNmNo8ZDdqawC1tiHohD33I7wk0p41eTRPCJlteK66Tge2BrWieUANwN7BGH+ueDTw/yewks4BdgR8OpMppZPCaBm1q3wnYrr01wKXAgYzxsHAJPGY0OO0YnftoxuLMS3IeTW/W5aPLtI91WwO4vqpubJvHXV4rrqq6HzgdOLGqFrXNC4CF7cD4d0yw7o3Ae9v1LwIuqKqvD7rmrnnneklagSXZEjiiqrae7lq0/GsH1V8A7FJVV053PTORPV6StIJKsjdwPPD+6a5Fy78kWwC/ohk4b+gahz1ekiRJHbHHS5IkqSMGL0mSpI4YvCRJkjpi8JI0cEkem+SE9lYYv0jy7SSbT3ddSyLJLkkuS3J6O318kgUTXR6/FPu6Z1lvU9LMsNze+VXS8qF9ft/JwLFV9bq2bS7NI0V+uZTbntVzr6BB2wv4h6o6PcljgedWVd83DU2yUlUtHFx5kpYH9nhJGrQX0jzu43OjDVV1YVWdlcbHk1yS5OIkrwVI8oIko89lI8ln2kcikeSaJB9I8iNglyT7tr1oC5Kc0C6zWpKjkpyb5OdJdmzbn5zknCQXtstvtnixSXZta7kkyUfbtg8AzwM+l+TjwPeBddrt/HWSTZN8N8n5Sc5K8qR2vWOSHNr2kn10guU2SfLTtt4PLftfgaSZwh4vSYP2FJqH2o5lZ5pnTW4JzAbOTXJmH9v8Y1U9DyDJDcAmVXVfkke38/8Z+EFV7dm2nZPkf4C9gU9W1XFJVgFm9W60fRD0R4FnAncA30/yyqr6YJJtgXdV1Xntg8e/OfqsuiSnAXtX1ZVJngX8B7Btu9nNgRdX1aIJlvsk8Nmq+nySfZA0tAxekqbT84Dj29OFNyf5Ic0z3u6aZL0v9bxfAByX5GvA19q2lwCvSPKudvoRwBzgp8A/J9kQ+OoYN3ncCjijqm4FSHIc8Dc92/0LSVYHnguc1JxVBeDhPYuc1IauiZbbBnhV+/4LNOFP0hAyeEkatEtpHq48lozTvpCHDoV4xGLz7+15/3c04egVwP9L8uR2u6+qqisWW++yJGe363wvyd9X1Q/6qGciDwPuHO39GsO9fS7n3aylFYBjvCQN2g+Ahyd582hDkq2SPB84E3htkllJRmgC1DnAtcAWSR6eZE3gRWNtuH0u3EZVdTrwbuDRwOrA94C3tQP7SfL09vXxwFVV9SngFOBpi23ybOD5SWYnmQXsCvxwog9XVXcBVyfZpd1H2ucfLslyPwZe177fbaL9SVq+GbwkDVQ1zyXbCdiuvZ3EpcCBwA00VzsuAC6iCWjvrqqbquo3wIntvOOAn4+z+VnAfye5uF3msKq6E/gQsDKwIMkl7TTAa4FLklwIPAn4/GK13gi8Fzi9remCqvp6Hx9zN2CvJBfR9PDtuITL7Qfsk+RcYM0+9idpOeWzGiVJkjpij5ckSVJHDF6SJEkdMXhJkiR1xOAlSZLUEYOXJElSRwxekiRJHTF4SZIkdcTgJUmS1JH/D/ARIRNyqbapAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np \n",
    "import matplotlib.pyplot as plt  \n",
    "  \n",
    "   \n",
    "# creating the dataset \n",
    "data = {'C':20, 'C++':15, 'Java':30,  \n",
    "        'Python':35} \n",
    "courses = list(data.keys()) \n",
    "values = list(data.values()) \n",
    "   \n",
    "fig = plt.figure(figsize = (10, 5)) \n",
    "  \n",
    "# creating the bar plot \n",
    "plt.bar(courses, values, color ='maroon',  \n",
    "        width = 0.4) \n",
    "  \n",
    "plt.xlabel(\"Courses offered\") \n",
    "plt.ylabel(\"No. of students enrolled\") \n",
    "plt.title(\"Students enrolled in different courses\") \n",
    "plt.show() \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
